Enhancement Proposal for Adding HTTP header to HAProxy Without Customizing haproxy.config Template (#1296)
openshift/enhancements: enhancements/ingress/set-replace-append-delete-deny-headers.md

  - "@gcs278"
  - "@Miciah"
api-approvers:
  - "@JoelSpeed"
  - "@deads2k"
  - "@frobware"
  - "@gcs278"
  - "@Miciah"
creation-date: 2022-12-02
last-updated: 2022-12-14
tracking-link:
  - "https://issues.redhat.com/browse/NE-982"
  - "https://issues.redhat.com/browse/RFE-464"
status: TODO
see-also:
  - ""
replaces:
superseded-by:
---

# Set, Delete, Deny HTTP Response headers

## Summary

This enhancement extends the IngressController API and Route API to allow the user to set or delete
HTTP response headers. Using the new API, users may specify a list of
HTTP response headers either via IngressController API or Route API that should be set or deleted.
Also, this enhancement creates an immutable(hard-coded) list of HTTP response headers which can't be set or deleted.

## Motivation

Certain HTTP response headers may need to be set, deleted for compliance reasons or other reasons. In OpenShift 3, cluster 
administrators could use a custom HAProxy configuration template to set or delete headers, but using a custom
template is not permitted in OpenShift 4.
Certain HTTP response headers like HSTS (Strict-Transport-Security) need to be denied as we already provide a way to set them

> via route annotation.

Reviewer: Would it not be better to provide a concrete API and deprecate the annotation API?

Author: Deprecating will be big job to do for now. We can try to achieve this in the next release.

Reviewer: I wonder is this still true, given this is going into 4.14, do we want to think about this deprecation now?

Reviewer 2: The HSTS annotation predates our era of learning that annotations were a mistake (3.7 I think https://docs.openshift.com/container-platform/3.7/admin_guide/managing_networking.html#admin-guide-enabling-hsts).  Since then we learned that the migration path from "temporary" annotation to formal schema basically never happens because never have the appetite to break consumers.

For types we own the schema for, annotation-based APIs should never be accepted.  Annotations
1. suffer from squatting: people can set the value before we add validation
2. suffer from lack of discoverability: we don't have a schema to inform generic editors
3. suffer from limitations of validation: certain kinds of validators aren't allowed on annotations, so hooks are more frequently used instead
4. suffer from low expressibility: we cannot extend future fields under a parent

For types we do not own schemas for, we can approve annotation-based APIs, but only when we can provide clear feedback for failures to the parties that can fix the problems.

### User Stories

1. As a cluster administrator, I need to configure an IngressController to set/delete certain HTTP response headers in order to
   satisfy compliance requirements at global level to send/not to send them to the client which has called any route from IngressController's domain.

> 2. As a user, I need to configure Route to set/delete HTTP response headers to send/not send them to the client which has called the route.

Reviewer: This user story is missing the why, you've told me who and what they want to do, but why is this important?

Author: there may be a variety of reasons.

### Goals

1. Enable the cluster administrator to specify HTTP response headers in the IngressController CR spec that should be set or 
   deleted at the global level, i.e. for all routes.
2. Enable the user to specify HTTP response headers in the Route spec (when it is not a passthrough Route) for their application
   that should be set or deleted when forwarding a response to the client.
3. Define an immutable/hard-coded deny list of headers which are not allowed to be set or deleted.
   Eg- we should prohibit specifying strict-transport-security using the new API, for the following reasons:

>    HSTS is already configurable through a route annotation.

Reviewer: IMO, as an API reviewer, I would suggest this isn't a good reason here. We should move away from annotation based APIs where possible

Author: Yes. But, removing this now will be a big thing to do. We can target this may be in the next release.

Reviewer 3: It's not just configurable through a route annotation, and that is a better reason why.  There is a whole API for configuring HSTS, at an admin level.

   We already have an API for restricting HSTS in the cluster config.

>    Adding a second way to configure HSTS would significantly complicate this enhancement.

Reviewer: Is it not just setting a particular header key and value?

Author: HSTS can be set via route annotation. However, the policy as to what can be set on a route annotation can be defined on the ingress config object.

Reviewer 4: > Is it not just setting a particular header key and value?

The `strict-transport-security` header (AKA HSTS) has wider ranging implications than the typical HTTP header.  We considered allowing HSTS to be configured on the IngressController in the past; see [the global-options-enable-hsts EP](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/global-options-enable-hsts.md), which was never implemented.  However, HSTS isn't neatly scoped:  Route A's HSTS policy can affect Route B, even if the routes are in different namespaces or belong to different IngressControllers, if Route B's host name is a subdomain of Route A's.  Additionally, specifying HSTS on an IngressController can cause a route manifest that behaves as expected on one cluster to behave surprisingly and problematically on another cluster.  For these reasons, we decided not to implement the global-options-enable-hsts EP and instead wrote and implemented [the global-admission-hsts EP](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/global-admission-hsts.md), which defines a cluster-scoped API, using which the cluster admin can specify HSTS parameters that may or must be specified on routes for a given domain: <https://github.com/openshift/api/blob/6ba31fa438f20d6c822ef47ad3d771309cb5216b/config/v1/types_ingress.go#L64-L87>.  Based on this experience, we wouldn't want to allow configuring HSTS on the IngressController using this new API.

Reviewer: Thanks Miciah, that's a really good justification for not allowing this, can we make sure this enhancement is updated to cover this so that future readers have that same context and don't ask the same question again

Author: done

### Non-Goals

1. Validation that each specified value is valid for the respective header (e.g. require that either "DENY" or "SAMEORIGIN" is 
   specified for [the `x-frame-options` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options)).

> 2. Setting load balancing algorithm using header values.
>    This request is orthogonal to customizing HTTP headers. It is out of scope for this enhancement and should be handled
>    with a seprate RFE.

Reviewer 5: ```suggestion
2. Setting load balancing algorithm using header values. The HAproxy configuration provides a [load balancing algorithm](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-balance) which relies on the check against the incoming request's header values. Even though this enhancement adds the header manipulation capability it doesn't aim at providing (or enhancing) the load balancing API for the IngressController or Route.  Therefore such a load balancing algotrithm is out of scope for this enhancement and should be handled with a separate RFE.
```

Author: done

3. Configuring buffer sizes by referring headers. Eg- tune.bufsize and tune.maxrewrite
4. Modifying headers for a passthrough route.

> 5. [Replace part of a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-part-of-a-header-by-using-a-regular-expression) 
> 6. [Replace a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-a-header-by-using-a-regular-expression)

Reviewer 5: ```suggestion
5. Providing dedicated API knobs for the [appending](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response%20add-header) or [replacing](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response%20replace-header) headers. The headers will be modified according to the [set-header HAProxy configuration stanza](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response%20set-header): add header if it doesn't exist and remove before adding if it does.
```

Author: done

    headersManipulation:
      response: 
      - name: X-Frame-Options
        action: Set
        set:
          value: DENY
      - name: X-XSS-Protection
        action: Delete
```
2. Route owner wants to set/delete HTTP response headers for only a specific route.
   Users edits/creates the Route having Spec as follows -
```yaml
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  labels:
    type: sharded 
  name: hello-openshift-edge
  namespace: hello-openshift
spec:
  subdomain: hello-openshift 
  tls:
    termination: edge
  to:
    kind: Service
    name: hello-openshift
  httpHeaders:
    headersManipulation:
      response:
      - name: X-Frame-Options
        action: Set
        set:
          value: DENY
      - name: X-XSS-Protection
        action: Delete
```

#### Variation [optional]

### API Extensions

A) Set/Delete HTTP response headers in the frontend sections of haproxy.config via IngressController's Spec.
The existing Ingress Controller API `HTTPHeaders *IngressControllerHTTPHeaders` is extended by adding a field for the header manipulations.

```go
// IngressControllerSpec is the specification of the desired behavior of the
// IngressController.

> type IngressControllerSpec struct {

Reviewer: Is this API up to date with the PR? I'm not going to review thoroughly as we can do that in the API PR, but there are a few comments I'd like to highlight to other reviewers

Author: yes it is up to date

type RouteHTTPHeaders struct {
    // headersManipulation specifies options for set/delete headers.
	// +kubebuilder:validation:Optional
	HeadersManipulation RouteHTTPHeadersManipulation `json:"headersManipulation,omitempty" protobuf:"bytes,1,opt,name=headersManipulation"`
}
```

```go
// RouteHTTPHeadersManipulation defines whether we want to set Response or Request Headers.
// Currently set/delete of HTTP Response Headers is available only.
type RouteHTTPHeadersManipulation struct {
    // response is a list of HTTP Headers which either need to set or deleted.
    Response []RouteSetOrDeleteHTTPHeader `json:"response,omitempty" protobuf:"bytes,1,opt,name=response"`
}
```

```go
// RouteSetOrDeleteHTTPHeader specifies options for set/delete a header.
type RouteSetOrDeleteHTTPHeader struct {
    // name specifies a header name to be set or deleted.  Its value must be a valid HTTP header.
    // name as defined in RFC 2616 section 4.2.
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=256
    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"

>     Name string `json:"name" protobuf:"bytes,1,opt,name=name"`

Reviewer: What would happen if I were to create multiple/conflicting entries with the same name in the list of headers?

Author: The latest value of the header in the list is set.

Reviewer: You need to make the slice for `request` and `response` into a `+listType=map` with `+listMapKey=name` to prevent duplicate entries with the same name, that will be less confusing for users since the API server will only persist a single value for each map key

Author: api has been updated.

    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=4096
    Value string `json:"value" protobuf:"bytes,2,opt,name=value"`
}
```

### Implementation Details/Notes/Constraints
A) Setting HTTP headers by cluster admin via IngressController's Spec.

Implementing this enhancement requires changes in the following repositories:

* openshift/api
* openshift/cluster-ingress-operator
* openshift/router

The router configures HAProxy using a configuration template.  The template uses
environment variables as input parameters.  The enhancement adds a new
environment variable, `ROUTER_HTTP_HEADERS`, which specifies the headers to set or deleted in the
following frontend sections of haproxy.config: `public`, `fe_no_sni`,`fe_sni`.
The ingress operator sets this variable based on the IngressController's
`spec.httpHeaders.headersManipulation.response[*].set` or `spec.httpHeaders.headersManipulation.response[*].delete` field values. 

The value of `ROUTER_HTTP_HEADERS` is a comma-delimited list of percent encoded  triplets of
the form `name:value:action` for action `Set` and `name:action` pairs for action `Delete`.

> The template translates these values into the appropriate `http-response set-header` stanzas.

Reviewer 4: What do we do with formats?  For example, what do we do with the following (inspired by <http://cbonte.github.io/haproxy-dconv/2.2/configuration.html#7.3.2-date>)?
```yaml
spec:
  httpHeaders:
    headersManipulation:
      response: 
      - name: expires
        action: Set
        set:
          value: "%[date(3600),http_date]"
```
Do we allow the format through, or do we we escape it?

Author: it will work as percent encoding will encode it %25%5Bdate%283600%29%2Chttp_date%5D
%[date(3600),http_date]
and again decode it %[date(3600),http_date]

Reviewer 4: All right.  We might be able to reuse the `IngressControllerHTTPUniqueIdHeaderPolicy` validation: https://github.com/openshift/api/blob/a7b1b96510144464f4b1123b7750894e0a2d95b7/operator/v1/types_ingress.go#L1268-L1280

We'll probably need to do some additional validation in the operator and router (e.g. verify that the format doesn't contain an invalid converter).

Author: For setting dynamic header values for http response. Allowed sample fetchers are res.hdr, ssl_c_der. Allowed converters are lower, base64.
  For setting dynamic header values for http request. Allowed sample fetchers are req.hdr, ssl_c_der. Allowed converters are lower, base64.

```go
// ServiceAliasConfig is a route for a service.  Uniquely identified by host + path.
type ServiceAliasConfig struct {
    ...
	// For route based setting of headers
	SetHTTPResponseHeaders              []SetHTTPHeader
	DeleteHTTPResponseHeaders           []DeleteHTTPHeader
}
```
```go
// SetHTTPHeader specifies an HTTP header that should be set.
type SetHTTPHeader struct {
	// Name specifies an HTTP header name.
	Name string

	// Value specifies the header value.
	Value string

	// Action specifies the action to be performed.
	Action string
}

type DeleteHTTPHeader struct {
	// Name specifies an HTTP header name.
	Name string
}
```

In this case the ordering of actions will be maintained as we will be having access to the RouteSpec inside router. So, we will run a loop
on either `spec.httpHeaders.headersManipulation.response[*].set` or `spec.httpHeaders.headersManipulation.response[*].delete` which will have the desired
sequence.

> C) We will be creating an immutable/harcoded deny list of HTTP Response headers in the [library-go](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141) 

Reviewer: If you do this, you will need to make sure that is documented in the godoc for this new API, I'm not sure I agree with the approach though, how many different headers are going to be blocked?

Author: For now we are blocking the HSTS header only as we already have a way to set this via route annotation.

  We need confirmation from the PM regarding this.

- Would we be able to set load balancing alogorithm based on headers ?

  Answer: No, this out of scope for this enhancement so it will be considered as a non-goal.

- Would we be able to delete or append headers  ?

  Answer: Yes, this enhancement will cover deletion and but not appending of HTTP Response Headers as per the [epic](https://issues.redhat.com/browse/RFE-464). 

- Would some headers be blocked for setting and deleting ?

  Answer: Yes, this enhancement won't allow the user to set or delete HSTS headers. We already a way to set this which is via route 
  annotation. Providing one more way to add this can cause issues.

- Would we be able to [Replace a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-a-header-by-using-a-regular-expression)?
  Answer: No, this is out of scope for this enhancement.

- Would we be able to [Replace part of a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-part-of-a-header-by-using-a-regular-expression) ?

>   Answer:  No, this is out of scope for this enhancement.

Reviewer: Could these be future extensions or do you not see this ever being supported?

Author: The epic does not demand it so we are not implementing this in this release. However the API can be extended by adding  a replace action.

Reviewer: How much extra work would it be to implement this now, in the initial implementation, vs later?

Author: not sure.

> 

Reviewer 5: One of the goals of this EP:
>  Define an immutable/hard-coded deny list of headers which are not allowed to be set or deleted.

This would forbid certain headers from being set/delete via the new API. However the use case you described is a little more complicated. `Cache-Control: private` is set by HAProxy's `cookie nocache` command which is needed for the [right sticky session implementation](https://www.haproxy.com/blog/enable-sticky-sessions-in-haproxy/#implement-sticky-sessions-with-a-cookie). The OpenShift IngressController follows this recommendation. So, since `Cache-Control: private` is set outside of this API, the previously mentioned deny list won't forbid it.

Also, this API will give you the possibility to remove this header and add it with the values you want unless we decide to forbid to do for `Cache-Control` header via the deny list (will be decided later).

> 

Reviewer 4: > Also, this API will give you the possibility to remove this header and add it with the values you want unless we decide to forbid to do so for the `Cache-Control` header via the deny list (will be decided later).

@alebedev87, @miheer, is one of you working on reaching that decision?

> 

Reviewer 4: It would be better to resolve the matter before we move on from the EP to the implementation.  At the very least, the EP should note that the `cache-control` header may be on the deny list.  

By the way, are you saying that we might need two deny lists: one for the IngressController API and one for the Route API?

> 

Reviewer 5: > By the way, are you saying that we might need two deny lists: one for the IngressController API and one for the Route API?

Not really. I just meant that it's not quite clear how HAProxy config execution will behave. Example:
```
backend be_secure:openshift-console:console
  http-response del-header Cache-Control
  cookie 1e2670d92730b515ce3a1bb65da45062 nocache ...
```
Knowing that `cookie` command will set `Cache-Control` header, will we still have it at end of the backend processing?

However we [know](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response) that the `frontend` section will be processed after `backend`, so adding `http-response del-header Cache-Control` will remove the header.

> 

Author: @Miciah black listing  of headers by users is a not a part of this epic. It will increase the scope.

> 

Author: if we want a deny list it will be like a new feature. We will need to revise the API as well.

> 

Author: For now as you rightly mentioned we can add cache-control in library-go code.

> 

Reviewer 6: Many applications serve mixed content: static assets (images, js, css) and dynamic content (html, json). They usually want to cache the static assets on the client and do not care about session stickiness for those requests. Only the dynamic requests need to stick to the same server (session sticky). 

Most applications will set something like this for static assets:
```
cache-control: max-age=7200, must-revalidate, public
```

But the OpenShift router (HAProxy) will add `cache-control: private` which results in both headers in the response:

```
cache-control: private 
cache-control: max-age=7200, must-revalidate, public
```

This leads to issues with caching: CDNs like Cloudflare will not cache your files. 

The best approach for users (developers) here would be to only add the `cache-control: private` if there is no `cache-control: public` already present in the response.

> 

Reviewer 4: > Many applications serve mixed content: static assets (images, js, css) and dynamic content (html, json). They usually want to cache the static assets on the client and do not care about session stickiness for those requests.

A common technique is to put static assets on a different virtual host.  In OpenShift terms, that means using a distinct route for static assets.  This route can be annotated with `haproxy.router.openshift.io/disable_cookies=true` to prevent setting the cookie and thus prevent setting `cache-control`.  Would that solution suffice?

> 

Reviewer 3: Don't forget that we also hard-code or explicitly configure some headers within the HAProxy template, e.g. https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L337-L348   and https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L537-L541, and some others.

> 

Author: Can you please check https://github.com/openshift/router/pull/438#discussion_r1170637633 ?

type RouteHTTPHeaderAction string

const (
    Set    RouteHTTPHeaderAction = "Set"
    Delete RouteHTTPHeaderAction = "Delete"
)
```

The RouteSetHTTPHeader type has fields for specifying a header name, header value and action.
```go
// RouteSetHTTPHeader specifies what value needs to be set on an HTTP header.
type RouteSetHTTPHeader struct {
    // Value specifies a header value.
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=4096
    Value string `json:"value" protobuf:"bytes,2,opt,name=value"`
}
```

RouteDeleteHTTPHeader type is an union member that is an empty struct.
```go
// RouteDeleteHTTPHeader is used specify delete action per header.
type RouteDeleteHTTPHeader struct {
}
```

### Implementation Details/Notes/Constraints
A) Setting HTTP headers by cluster admin via IngressController's Spec.

Implementing this enhancement requires changes in the following repositories:

* openshift/api
* openshift/cluster-ingress-operator
* openshift/router

The router configures HAProxy using a configuration template.  The template uses
environment variables as input parameters.  The enhancement adds a new

> environment variable, `ROUTER_HTTP_HEADERS`, which specify the headers to set or deleted in the

Author: done

following frontend sections of haproxy.config: `public`, `fe_no_sni`,`fe_sni`.
The ingress operator sets this variable based on the IngressController's
`spec.httpHeaders.headersManipulation.response[*].set` or `spec.httpHeaders.headersManipulation.response[*].delete` field values. 

The values of `ROUTER_HTTP_HEADERS` are comma-delimited lists of percent encoded values of

> the form `name:value:action` separated by comma for action `Set` and `name:action` for action `Delete`.

Author: done

The template translates these values into the appropriate `http-response set-header` stanzas.
For example, if `ROUTER_HTTP_HEADERS` has percent encoded value `X-Frame-Options%3A+DENY%3A+Set,X-XSS-Protection%3A+1%3Bmode%3Dblock%3A+Delete` 
of `X-Frame-Options: DENY:Set,X-XSS-Protection: 1;mode=block:Delete` then the template adds the following stanzas to

> the frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config:

Author: there are three so sections is right?

> the frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config:

Author: done

```conf
    http-response set-header X-Frame-Options: DENY
    http-response del-header X-XSS-Protection: 1;mode=block    
```

> Use of a single environment variable `ROUTER_HTTP_HEADERS` for bothe action `Set` and `Delete` will help preserve ordering.

Reviewer 5: ```suggestion
Use of a single environment variable `ROUTER_HTTP_HEADERS` for both actions `Set` and `Delete` will help to preserve ordering.
```

Author: done

   in order to satisfy compliance requirements at global level to send them to all clients.
2. As a user/devops/developer, I need to configure Route to set/append HTTP Response headers which need to  which 
   are mentioned above to send them to the client which has called the route.
3. To satisfy this use-case, the cluster administrator can set the IngressController's Spec 
   or the developer can set Router's `spec.httpDeleteHeaders` field to set(replace) or append the required headers.
4. As a cluster administrator, I need to configure an IngressController to delete certain HTTP Response headers which are:
   - X-XSS-Protection
   - X-Content-Type-Options
   - X-Frame-Options
     in order to satisfy compliance requirements at global level to not to send them to all clients.
5. As a user/devops/developer, I need to configure Route to delete HTTP Response headers which need to  which
      are mentioned above for not sending them to the client which has called the route.
6. To satisfy this use-case, the cluster administrator can set the IngressController's Spec
   or the developer can set the Router's Spec with `spec.httpDeleteHeaders` field to delete the required headers by 

>    mentioning the header names mentioned above.

Reviewer 4: Is there precedent for writing, "As a developer, I need to update ___ API" as a user story?  User stories are usually focused on end-user roles, such as cluster admin or project admin.  

I do like simplifying the user stories where possible and simplifying the list so that it enumerates one user story per list item.  It is helpful to provide the solution to the user story though.  When writing an EP, I put the user story as a heading or italicized statement followed by one or more paragraphs explaining how the EP addresses the user story.  

The user story should indicate who the end-user is (e.g. "cluster admin"), what the end user is trying to achieve (e.g. "delete the `x-server` response header in responses to client requests") and why ("so that I can avoid exposing information about internal systems to potentially malicious external actors").  Then you can explain how the EP solves that user story ("The cluster admin can configure the IngressController as shown in the following example yaml").

>    mentioning the header names mentioned above.

Reviewer 5: > Is there precedent for writing, "As a developer, I need to update ___ API" as a user story? User stories are usually focused on end-user roles, such as cluster admin or project admin.

I created such user stories myself, for technical debt for instance. I suppose it's in the interest of NetEdge engineer to define and implement the deny list. That's why I put Developer as the actor for the first user story.

> The user story should indicate who the end-user is (e.g. "cluster admin"), what the end user is trying to achieve (e.g. "delete the x-server response header in responses to client requests") and why ("so that I can avoid exposing information about internal systems to potentially malicious external actors"). Then you can explain how the EP solves that user story ("The cluster admin can configure the IngressController as shown in the following example yaml").

Isn't the Workflow chapter for this purpose?

>    mentioning the header names mentioned above.

Reviewer 4: > > Is there precedent for writing, "As a developer, I need to update ___ API" as a user story? User stories are usually focused on end-user roles, such as cluster admin or project admin.
> 
> I created such user stories myself, for technical debt for instance. I suppose it's in the interest of NetEdge engineer to define and implement the deny list. That's why I put Developer as the actor for the first user story.

I can see creating user stories in Jira for tech debt, and we sometimes note potential future follow-up work in the goals/non-goals, proposal, or implementation details/notes/constraints sections.  However, I still believe that user stories in the enhancement should be focused on end users.  

> > The user story should indicate who the end-user is (e.g. "cluster admin"), what the end user is trying to achieve (e.g. "delete the x-server response header in responses to client requests") and why ("so that I can avoid exposing information about internal systems to potentially malicious external actors"). Then you can explain how the EP solves that user story ("The cluster admin can configure the IngressController as shown in the following example yaml").
> 
> Isn't the Workflow chapter for this purpose?

I would use user stories for the high-level description of the problem ("As a ___, I want to ___, so that I can ___") and solution ("To accomplish ___, ___ can use the new ___ API/field/whatever as in the following example") and then document the detailed step-by-step procedure as a workflow.

>    mentioning the header names mentioned above.

Author: done

### Goals

> 1. Enable the cluster administrator to specify HTTP response headers in the IngressController CR spec that should get set or appended at the global
>    level, i.e. for all routes.
> 2. Enable the route owner to specify HTTP response headers in the Route spec (when it is not a passthrough Route) for their application
>    that should get set or appended when forwarding a response to the client.
> 3. Enable the cluster administrator to delete the HTTP response headers via IngressController CR.
> 4. Enable the developer to delete the HTTP response headers via RouteSpec.
> 5. This EP will set an immutable/hardcoded deny list of headers not allowed to set/append or deleted.
>    Eg- we should prohibit specifying strict-transport-security using the new API, for the following reasons:
>    HSTS is already configurable through a route annotation.
>    We already have an API for restricting HSTS in the cluster config.
>    Adding a second way to configure HSTS would significantly complicate this enhancement.

Reviewer 5: ```suggestion
1. Enable the cluster administrator to specify HTTP response headers in the IngressController CR spec that should be set, appended or deleted at the global level, i.e. for all routes.
2. Enable the route owner to specify HTTP response headers in the Route spec (when it is not a passthrough Route) for their application
   that should be set,  appended or deleted when forwarding a response to the client.
3. This EP will set an immutable/hard-coded deny list of headers which are allowed to set, appended or deleted.
   Eg- we should prohibit specifying strict-transport-security using the new API, for the following reasons:
   HSTS is already configurable through a route annotation.
   We already have an API for restricting HSTS in the cluster config.
   Adding a second way to configure HSTS would significantly complicate this enhancement.
```

Author: done

### Non-Goals

1. Validation that each specified value is valid for the respective header (e.g. require that either "DENY" or "SAMEORIGIN" is 
   specified for [the `x-frame-options` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options)).
2. Setting load balancing algorithm using header values.
   This request is orthogonal to customizing HTTP headers. It is out of scope for this enhancement and should be handled
   with a seprate RFE.
3. Configuring buffer sizes. Eg- tune.bufsize and tune.maxrewrite
4. Modifying headers for a passthrough route.

## Proposal

### Workflow Description

**Cluster admin** is a human user responsible for deploying a
cluster.

> **Users/Devops/Developers** is a human user responsible for developing and

Reviewer 5: I think that the end user of the API is a different actor from the developer which designs the API. The difference between them can be seen, for instance, in the part where the deny list is defined. The user actor cannot influence this list in any form, while developer has to be define it.
Also what did you mean by the devops actor? How it's different from the user of the API?

Author: I meant application developer

Reviewer 5: I see, then maybe this can be condensed. `User/DevOps/Application Developer` may simply become `User`. And I think that `Developer` as an actor representing NetworkEdge Engineer is still needed, for the deny list use case (user story and other mentions).

Author: changed

```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpDeleteHeaders:
    response:
    - name: X-Frame-Options
    - name: X-XSS-Protection
```
7. Users/Devops/Developers want to delete HTTP response header which need to be sent only to the client hitting
   the route with the HTTP Response Header values.
8. Users/Devops/Developers edits/creates the Route having Spec as follows -
```yaml
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  labels:
    type: sharded 
  name: hello-openshift-edge
  namespace: hello-openshift
spec:
  subdomain: hello-openshift 
  tls:
    termination: edge
  to:
    kind: Service
    name: hello-openshift
  httpDeleteHeaders:
    response:
    - name: X-Frame-Options
    - name: X-XSS-Protection
```
#### Variation [optional]

### API Extensions

A) Set(Replace)/Append HTTP headers in the global section of haproxy.config via IngressController's Spec.
The Ingress Controller API is extended by adding an optional `httpSetHeaders` with type 
`IngressControllerSetHTTPHeaders` to `IngressControllerSpec`.

```go

> type IngressControllerSpec struct {

Author: // httpHeaders defines policy for HTTP headers.
	//
	// If this field is empty, the default values are used.
	//
	// +optional
	HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`

The goal of this seems to be different i.e to set policies. But we can add the new field under this. WDYT @Miciah ?

> type IngressControllerSpec struct {

Reviewer 4: Using `IngressControllerHTTPHeaders` makes sense to me.  Can we do that and still keep the new IngressController API and the new Route API consistent?

> type IngressControllerSpec struct {

Author: @Miciah https://github.com/openshift/api/pull/1364

> type IngressControllerSpec struct {

Author: done

  //
  // If this field is empty, no response headers are set(replace) or append.
  //
  // +nullable
  // +optional
  Response []IngressControllerSetHTTPHeader `json:"response,omitempty"`
}
```

The IngressControllerSetHTTPHeader type has fields for specifying a header name, header value and action.
```go
type IngressControllerSetHTTPHeader struct{
    // name specifies a header name.  Its value must be a valid HTTP header 
    // name as defined in RFC 2616 section 4.2.
    //	
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=256
    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$" 
    // +required 
    Name string `json:"name"`

    // Value specifies a header value.
    // +kubebuilder:validation:Required 
    // +required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=4096
    Value string `json:"value"`

    // Action specifies a header value.
    // +kubebuilder:validation:Required
    // +required
    // +kubebuilder:validation:Enum=Append;Replace
    Action IngressControllerHTTPHeaderAction `json:"action"`
}
```

> We will be using the exisiting API:

Reviewer 4: This sentence needs to be fixed.

Author: done

```go
// IngressControllerHTTPHeaderAction is a policy for setting HTTP headers.
// +kubebuilder:validation:Enum=Append;SetOrReplace
type IngressControllerHTTPHeaderAction string

const (
  // AppendHTTPHeader appends the header, preserving any existing header.
  AppendHTTPHeader IngressControllerHTTPHeaderAction = "Append"
  // SetOrReplaceHTTPHeader sets the header, removing any existing header.
  SetOrReplaceHTTPHeader IngressControllerHTTPHeaderAction = "SetOrReplace"

> )

Reviewer 5: Yeah, I think you have the point about the `Value` field. It's probably better to be validated by the OpenAPI, this may be clearer for the end user.
However, what about the actions? How do you plan to use them in the cluster-ingress-operator or router?

> )

Reviewer 4: Consider using a [discriminated union](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions) where `IngressControllerHTTPHeaderAction` would be the discriminator, "Delete" would be an empty union member, and "Set" would be union member with a "Value" field.  "Replace" and "Append" could be expressed as separate union members or as a field of the "Set" union member.

> )

Reviewer 5: I can see the delete as a union member:
```go
// +union
type IngressControllerHTTPHeader struct {
  // +unionDiscriminator
  // +kubebuilder:validation:Required
  Action string `json:"action,omitempty"`
  // +optional.
  Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`
  // +optional.
  Delete *IngressControllerDeleteHTTPHeader `json:"delete,omitempty"`
}
```
```yaml
apiVersion: operator.openshift.io/v1                                                                                                         
kind: IngressController
metadata:
  name: default
spec:
  httpHeaders:
    headerManipultation:
      response:
      - action: "Set"
        name: "X-Frame-Options"
        value: "DENY"
      - action: "Delete"
        name: "X-Powered-By"
```
but I cannot quite understand how it can be an empty union member.
Does it mean that only `Delete` action will be specified?

> )

Author: For Action we can specify 

> Consider using a [discriminated union](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions) where `IngressControllerHTTPHeaderAction` would be the discriminator, "Delete" would be an empty union member, and "Set" would be union member with a "Value" field. "Replace" and "Append" could be expressed as separate union members or as a field of the "Set" union member.

yeah it won't empty. IngressControllerDeleteHTTPHeader will need header name which needs to be passed.

> )

Reviewer 5: Thanks @Miciah for clarifying.
I'm almost fine with the discriminated union approach. It just feels a little verbose, especially on this part:
```yaml
- action: "Set"
  name: "X-Frame-Options"
  set:  <-- this part
    value: "DENY"
```

> )

Reviewer 4: > I'm almost fine with the discriminated union approach. It just feels a little verbose

True, but it's a standard pattern and makes it clear which fields apply to which action.

> )

Reviewer 5: Sorry to bring this back but I don't think I'm 100% convinced that the discriminated union approach is better than the dedicated Set and Delete fields: 
```yaml
apiVersion: operator.openshift.io/v1                                                                                                         
kind: IngressController
metadata:
  name: default
spec:
  httpHeaders:
    headerManipultation:
      response:
        set:
        - name: "X-Frame-Options"
          value: "DENY"
        delete:
        - name: "X-Powered-By"
```

The API maybe a little less redundant for the discriminated union (`Name` field is common, no `DeleteHeader` type is needed) but is it more convenient to the end user? IMO it feels a little more verbose than the variant I showed above.

> )

Reviewer 5: The discriminated union approach allows the API users to create an ordered list of the header manipulation instructions. This maybe more explicit expression of their intent as the dedicated set/delete fields approach would always raise the question in which order the two lists need to be translated into HAProxy configuration.

> )

Reviewer 4: > Btw don't we loose some validation using `Name` as an empty union member ([same question](https://github.com/openshift/enhancements/pull/1296#discussion_r1064629179))? it has to be optional while we need it required, no?

For the record, I believe I answered the question here: https://github.com/openshift/enhancements/pull/1296#discussion_r1064668220

> `Name` should not be a union member. You can use a inline field for the union type. See `IngressControllerCaptureHTTPCookie` ([source](https://github.com/openshift/api/blob/5ad311dde2264b659a256a17c27f1836fb50e85d/operator/v1/types_ingress.go#L1100-L1117)) for an example.

Please let me know if that doesn't satisfactorily answer your question!

  // name specifies a header name.  Its value must be a valid HTTP header
  // name as defined in RFC 2616 section 4.2.
  //	
  // +kubebuilder:validation:Required
  // +kubebuilder:validation:MinLength=1   
  // +kubebuilder:validation:MaxLength=256   
  // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"	
  // +required
  Name string `json:"name"`
}
```

C) Setting HTTP headers in the backend section `Plain http backend or backend with TLS terminated at the edge or a 
secure backend with re-encryption.` of the haproxy.config via Route's Spec.
The Route API's Spec is extended by adding optional `httpSetHeaders` of type `RouteSetHTTPHeaders` :
```go
type RouteSpec struct {
	...
	
	// httpSetHeaders defines HTTP headers that should be set/append.
	// If this field is empty, no headers are added.
	//
	// Note that this option only applies to cleartext HTTP connections
	// and to secure HTTP connections for which the ingress controller
	// terminates encryption (that is, edge-terminated or reencrypt
	// connections).  Headers cannot be set/append for TLS passthrough
	// connections.
	//
	// +kubebuilder:validation:Optional
	// +optional

> 	HTTPSetHeaders RouteSetHTTPHeaders `json:"httpSetHeaders,omitempty"`

Author: done

  // name as defined in RFC 2616 section 4.2.
  //
  // +kubebuilder:validation:Required
  // +kubebuilder:validation:MinLength=1
  // +kubebuilder:validation:MaxLength=256
  // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
  // +required 
  Name string `json:"name"`

  // Value specifies a header value.
  // +kubebuilder:validation:Required
  // +required
  // +kubebuilder:validation:MinLength=1
  // +kubebuilder:validation:MaxLength=4096
  Value string `json:"value"`

  // Action specifies a header value.
  // +kubebuilder:validation:Required
  // +required
  // +kubebuilder:validation:Enum=Append;Replace
  Action RouteHTTPHeaderAction `json:"action"`
}
```

```go
// RouteHTTPHeaderAction is a policy for setting, replacing or appending HTTP headers.
//
// +kubebuilder:validation:Enum=Append;SetOrReplace
type RouteHTTPHeaderAction string

const (
	// AppendHTTPHeaderPolicy appends the header, preserving any existing header.
	AppendHTTPHeader RouteHTTPHeaderPolicy = "Append"
	// ReplaceHTTPHeaderPolicy sets the header, removing any existing header.
	ReplaceHTTPHeaderPolicy RouteHTTPHeaderPolicy = "SetOrReplace"

> )

Author: changed

    //
    // If this field is empty, the default values are used.
    //
    // +optional
    HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`
}
```

```go

// IngressControllerHeadersManipulation specifies options for set/replace/append/delete headers.
type IngressControllerHeadersManipulation struct {
    Response []IngressControllerSetReplaceAppendDeleteHTTPHeaders `json:"response,omitempty"`
}

```

```go
// IngressControllerHeadersManipulation specifies options for set/replace/append/delete headers.
// +union
type IngressControllerSetReplaceAppendDeleteHTTPHeaders struct {
    // action lets you specify headers for Set or Delete.
    // +unionDiscriminator
    // +kubebuilder:validation:Enum:=Set;Delete 
	//+kubebuilder:validation:Required
    Action IngressControllerSetHTTPHeaderAction `json:"action,omitempty"`

    // name specifies a header name.  Its value must be a valid HTTP header
    // name as defined in RFC 2616 section 4.2.
    // +kubebuilder:validation:Optional
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=256
    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"

>     // +optional

Reviewer 5: Can empty union member be required? Because we need it required.

Reviewer 4: `Name` should not be a union member.  You can use a inline field for the union type.  See `IngressControllerCaptureHTTPCookie` ([source](https://github.com/openshift/api/blob/5ad311dde2264b659a256a17c27f1836fb50e85d/operator/v1/types_ingress.go#L1100-L1117)) for an example.

Author: do you mean something like this -
`    // headersManipulation specifies options for set/replace/append/delete headers.
   HeadersManipulation IngressControllerHeadersManipulation `json:"headersManipulation,omitempty"`
}

type IngressControllerHeadersManipulation struct{
   Response IngressControllerSetOrDeleteHeaders `json:"response,omitempty"`
}

// IngressControllerHeadersManipulation specifies options for set/replace/append/delete headers.
type IngressControllerSetOrDeleteHeaders struct {
   // name specifies a header name to be set(replace)/append/delete.  Its value must be a valid HTTP header.
   // name as defined in RFC 2616 section 4.2.
   // +kubebuilder:validation:Required
   // +kubebuilder:validation:MinLength=1
   // +kubebuilder:validation:MaxLength=256
   // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
   // +required
   Name *string `json:"name"`

   IngressControllerSetReplaceAppendDeleteHTTPHeaders `json:",inline"`
}

// IngressControllerSetReplaceAppendDeleteHTTPHeaders specifies fields to mention the headers and the action you want to perform on them.
// +union
type IngressControllerSetReplaceAppendDeleteHTTPHeaders struct {
   // action lets you specify headers for Set or Delete.
   // +unionDiscriminator
   // +kubebuilder:validation:Enum:=Set;Delete
   // +kubebuilder:validation:Required
   Action IngressControllerSetHTTPHeaderAction `json:"action,omitempty"`

   // Set defines HTTP headers that should be set or replaced or appended.
   // If this field is empty, no headers are added.
   //
   // Note that this option only applies to cleartext HTTP connections
   // and to secure HTTP connections for which the ingress controller
   // terminates encryption (that is, edge-terminated or reencrypt
   // connections).  Headers cannot be set(replace)/append for TLS passthrough
   // connections.
   //
   // +kubebuilder:validation:Optional
   // +optional
   Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`

   // delete defines HTTP headers that should be deleted.
   // If this field is empty, no headers are deleted.
   //
   // Note that this option only applies to cleartext HTTP connections
   // and to secure HTTP connections for which the ingress controller
   // terminates encryption (that is, edge-terminated or reencrypt
   // connections).  Headers cannot be set(replace)/append for TLS passthrough
   // connections.
   //
   // +kubebuilder:validation:Optional
   // +optional
   Delete *IngressControllerDeleteHTTPHeader `json:"delete,omitempty"`
}`

Author: The CR will look like this ->
~~~
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpHeaders:
    headersManipulation:
      response: 
      - name: X-Frame-Options
        action: Set
        set:
          value: DENY
          action: Replace
      - name: X-XSS-Protection
        action: Set
        set:
          value: 1;mode=block
          action: Replace
~~~

Author: @Miciah Can you please check this ?

Reviewer 4: Sorry for not responding to this comment earlier.  Do you still need me to check something?

Author: api changes are updated

    Name *string `json:"name"`

    // Set defines HTTP headers that should be set or replaced or appended.
    // If this field is empty, no headers are added.
    //
    // Note that this option only applies to cleartext HTTP connections
    // and to secure HTTP connections for which the ingress controller
    // terminates encryption (that is, edge-terminated or reencrypt
    // connections).  Headers cannot be set(replace)/append for TLS passthrough
    // connections.
    //
    // +kubebuilder:validation:Optional
    // +optional
    Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`

    // delete defines HTTP headers that should be deleted.
    // If this field is empty, no headers are deleted.
    //
    // Note that this option only applies to cleartext HTTP connections
    // and to secure HTTP connections for which the ingress controller
    // terminates encryption (that is, edge-terminated or reencrypt
    // connections).  Headers cannot be set(replace)/append for TLS passthrough
    // connections.
    //
    // +kubebuilder:validation:Optional
    // +optional

>     Delete *IngressControllerDeleteHTTPHeader `json:"delete,omitempty"`

Reviewer 5: Why `Delete` union member is needed? It doesn't have any data to carry.

Reviewer 4: It can be omitted but IMO it is clearer to include it.

Author: api changes are updated

    Delete IngressControllerSetHTTPHeaderAction = "Delete"
)
```

The IngressControllerSetHTTPHeader type has fields for specifying a header name, header value and action.
```go
type IngressControllerSetHTTPHeader struct{
	// Value specifies a header value.
    // +kubebuilder:validation:Required 
    // +required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=4096
    Value string `json:"value"`

    // Action specifies a header value.
    // +kubebuilder:validation:Required
    // +required
    // +kubebuilder:validation:Enum=Append;Replace
    Action IngressControllerHTTPHeaderAction `json:"action"`
}
```

This defines Actions to be perfomed on headers when action field is `Set` in the IngressControllerHTTPHeaders. 
```go
// IngressControllerHTTPHeaderAction is a policy for setting HTTP headers.
// +kubebuilder:validation:Enum=Append;Replace
type IngressControllerHTTPHeaderAction string

const (
    // AppendHTTPHeader appends the header, preserving any existing header.
    AppendHTTPHeader IngressControllerHTTPHeaderAction = "Append"
    // ReplaceHTTPHeader sets the header, removing any existing header.

>     ReplaceHTTPHeader IngressControllerHTTPHeaderAction = "Replace"

Reviewer 5: I still think that the name is a little misleading for the end user. `http-response set-header`which will be used for the replace action can not only replace a header but can add the header if it doesn't exist. We need to do something about it, either change the action name to reflect better what will happen or change the implementation.
I personally thought (maybe too radical though) of getting rid of the action type as a whole: this would map 1-1 with `http-response set-header` config command.

Reviewer 4: It would be helpful to have a table in the implementation details showing how each action maps to a stanza in `haproxy.config`.

Reviewer 5: As discussed during the design meeting (09/01/23): no explicit requirement about adding header or replacing by regexp was mentioned in RFE. We assume `http-response set-header` (add if doesn't exist replace if does) stanza is what corresponds the most to the customer expectations.

Author: done

    // Action specifies a header value.
    // +kubebuilder:validation:Required
    // +required
    // +kubebuilder:validation:Enum=Append;Replace
    Action RouteHTTPHeaderAction `json:"action"`
}
```

```go
// RouteHTTPHeaderAction is a policy for setting, replacing or appending HTTP headers.
//
// +kubebuilder:validation:Enum=Append;Replace
type RouteHTTPHeaderAction string

const (
	// AppendHTTPHeaderPolicy appends the header, preserving any existing header.
	AppendHTTPHeader RouteHTTPHeaderPolicy = "Append"
	// ReplaceHTTPHeaderPolicy sets the header, removing any existing header.
	ReplaceHTTPHeaderPolicy RouteHTTPHeaderPolicy = "Replace"
)
```

The RouteDeleteHTTPHeader is an empty union member .
```go
type RouteDeleteHTTPHeader struct{
}
```

### Implementation Details/Notes/Constraints
A) Setting HTTP headers by cluster admin via IngressController's Spec.

Implementing this enhancement requires changes in the following repositories:

* openshift/api
* openshift/cluster-ingress-operator
* openshift/router

The router configures HAProxy using a configuration template.  The template uses
environment variables as input parameters.  The enhancement adds a new

> environment variables, `ROUTER_SET_HTTP_HEADERS`, which specify the headers to set in the

Reviewer 5: I'm still confused about `ROUTER_SET_HEADERS` variable. Would it carry all the action types (`Append` and `Replace`)? How would you distinguish which headers need to be appended and which set?

Author: Yes.  We will append action along with `setheader name:setheader value: action`in the env var.

Reviewer 5: As per [the design discussion](https://github.com/openshift/enhancements/pull/1296#discussion_r1065042615) we won't have append and replace actions, so no need to make a difference.

Author: No we will need the Action to be passed in env var of router deployment.

Author: We will pass Set or Delete

Author: I think I get your point, you mean we can just check if ROUTER_SET_HTTP_HEADERS is set , if yes then the Action is Set.

  We need confirmation from the PM regarding this.

- Would we be able to set load balancing alogorithm based on headers ?

  Answer: No, this out of scope for this enhancement so it will be considered as a non-goal.

- Would we be able to delete or append headers  ?

  Answer: Yes, this enhancment will cover deletion and appending of HTTP Response Headers as it is a requirement of the epic.

- Would some headers be blocked for setting, deleting and appending ?

  Answer: Yes, this enhancement won't allow the user to set/replace, append or delete HSTS headers. We already a way to set this which is via route 
  annotation. Providing one more way to add this can cause issues.

- Would we be able to Replace a header by using a regular expression ?
  Answer: We need to check on this with the PM. [Replace a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-a-header-by-using-a-regular-expression)

- Would we be able to Replace part of a header by using a regular expression ?

>   Answer: We need to check on this with the PM. [Replace part of a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-part-of-a-header-by-using-a-regular-expression) 

Author: @mcurry-rh  @Deepthidharwar can you please check this ?

Author: done

---
title: set-replace-append-delete-deny-headers
authors:
  - "@miheer"
reviewers:
  - "@frobware"
  - "@gcs278"
  - "@Miciah"
approvers:
  - "@frobware"
  - "@gcs278"
  - "@Miciah"

> api-approvers:

Reviewer: This should include either myself or @deads2k as an API approver, I see there's an API PR so we can discuss the API there, but this shouldn't be merged until the API has been approved

Author: done

  - "@frobware"
  - "@gcs278"
  - "@Miciah"
creation-date: 2022-12-02
last-updated: 2022-12-14
tracking-link:
  - "https://issues.redhat.com/browse/NE-982"
  - "https://issues.redhat.com/browse/RFE-464"
status: TODO
see-also:
- ""
replaces:
superseded-by:
---

# Set, Delete, Deny HTTP Response headers

## Summary

This enhancement extends the IngressController API and Route API to allow the user to set or delete
HTTP response headers. Using the new API, users may specify a list of
HTTP response headers either via IngressController API or Route API that should be set or deleted.
Also, this enhancement creates an immutable(hard-coded) list of HTTP response headers which can't be set or deleted.

## Motivation

Certain HTTP response headers may need to be set, deleted for compliance reasons or other reasons. In OpenShift 3, cluster 
administrators could use a custom HAProxy configuration template to set or delete headers, but using a custom
template is not permitted in OpenShift 4.
Certain HTTP Response headers like HSTS(Strict-Transport-Security) need to denied as we already provide a way to set them

> via route annotation.

Reviewer 4: Also, you need add a space  before the open parenthesis, and is there a reason to capitalize "Response"?

> via route annotation.

Author: generally we capitalise it right ?

> via route annotation.

Reviewer 4: I don't see a reason to capitalize it generally.

> via route annotation.

Author: ok for now now I have kept HTTP. if you feel strongly then I will not capitalize it.

### User Stories

> 1. As a developer, I need to update IngressController and Route APIs to be be able to set/delete certain HTTP response headers.

Author: so basically it says I want to update the IngressController CR and Route Object to be be able to set/delete certain HTTP response headers.

> 1. As a developer, I need to update IngressController and Route APIs to be be able to set/delete certain HTTP response headers.

Author: it is not needed I will remove it.

> 1. As a developer, I need to update IngressController and Route APIs to be be able to set/delete certain HTTP response headers.

Author: I don't think deny list can come in a user story. Because user can't do anything.

   We already have an API for restricting HSTS in the cluster config.
   Adding a second way to configure HSTS would significantly complicate this enhancement.

### Non-Goals

1. Validation that each specified value is valid for the respective header (e.g. require that either "DENY" or "SAMEORIGIN" is 
   specified for [the `x-frame-options` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options)).
2. Setting load balancing algorithm using header values.
   This request is orthogonal to customizing HTTP headers. It is out of scope for this enhancement and should be handled
   with a seprate RFE.
3. Configuring buffer sizes by referring headers. Eg- tune.bufsize and tune.maxrewrite
4. Modifying headers for a passthrough route.
5. [Replace part of a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-part-of-a-header-by-using-a-regular-expression) 
6. [Replace a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-a-header-by-using-a-regular-expression)

## Proposal

> ### Workflow Description

Author: fixed

**Cluster admin** is a human user responsible for deploying a
cluster.

**Users** is a human user responsible for developing and
deploying an application in a cluster.

> 1. Cluster admin wants to set(replace)/append HTTP response headers at global level which need to be sent to all clients. 

Author: done

2. Cluster admin edits/creates the CR of the Ingress Controller having Spec as follows -
```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpHeaders:
    headersManipulation:
      response: 
      - name: X-Frame-Options
        action: Set
        set:
          value: DENY
      - name: X-XSS-Protection
        action: Set
        set:
          value: 1;mode=block
```

> 3. Users want to set(replace)/append HTTP response header which need to be sent only to the client hitting 
>    the route with the HTTP Response Header values.

Reviewer 4: The phrase "sent only to the client hitting the route" seems confusing to me.  I like the suggestion for the other use-case, where you replace "sent to all clients" with "for all the routes".  How about doing likewise here?  
```suggestion
3. Route owner wants to set HTTP response headers for only a specific route.
```

>    the route with the HTTP Response Header values.

Author: fixed

4. Users edits/creates the Route having Spec as follows -
```yaml
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  labels:
    type: sharded 
  name: hello-openshift-edge
  namespace: hello-openshift
spec:
  subdomain: hello-openshift 
  tls:
    termination: edge
  to:
    kind: Service
    name: hello-openshift
  httpHeaders:
    headersManipulation:
      response:
      - name: X-Frame-Options
        action: Set
        set:
          value: DENY
      - name: X-XSS-Protection
        action: Set
        set:
          value: 1;mode=block
```

> 5. Cluster admin wants to delete HTTP response headers at global level which need to be sent to all clients.

Author: fixed

6. Cluster admin edits/creates the CR of the Ingress Controller having Spec as follows -
```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
   httpHeaders:
     headersManipulation:
       response:
       - name: X-Frame-Options
         action: Delete
         delete:  
       - name: X-XSS-Protection
         action: Delete
         delete:
```
7. Users want to delete HTTP response header which need to be sent only to the client hitting

>    the route with the HTTP Response Header values.

Author: fixed

8. Users edits/creates the Route having Spec as follows -
```yaml
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  labels:
    type: sharded 
  name: hello-openshift-edge
  namespace: hello-openshift
spec:
  subdomain: hello-openshift 
  tls:
    termination: edge
  to:
    kind: Service
    name: hello-openshift
    httpHeaders:
  headersManipulation:
    response:
    - name: X-Frame-Options
      action: Delete

>       delete:

Reviewer 4: It isn't needed.

>       delete:

Author: api is fixed

    - name: X-XSS-Protection
      action: Delete
      delete:
```
#### Variation [optional]

### API Extensions

A) Set/Delete HTTP headers in the global section of haproxy.config via IngressController's Spec.
The existing Ingress Controller API `HTTPHeaders *IngressControllerHTTPHeaders` is extended by adding an field for action to be performed,

> and field for headers to be set/deleted.

Author: fixed

    // +optional
    HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`
}
```

```go

type IngressControllerHTTPHeaders struct {
	

> 	// +unionDiscriminator

Reviewer 5: For this work, `IngressControllerHTTPHeaders` needs to be set as a union (`+union`) which would break the API. You need a dedicated field for the header manipulation.

Author: changed

    HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`
}
```

```go
// IngressControllerHTTPHeaders specifies how the IngressController handles
// certain HTTP headers.
type IngressControllerHTTPHeaders struct { 

> 	// headersManipulation specifies options for set/delete headers.

Author: api has been updated

	// terminates encryption (that is, edge-terminated or reencrypt
	// connections).  Headers cannot be set(replace)/append for TLS passthrough
	// connections.
	//
	// +kubebuilder:validation:Optional
	// +optional

> 	Set IngressControllerSetHTTPHeaders `json:"set,omitempty"`

Reviewer 5: Note that all the union members [must be pointers](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#writing-a-union-in-go).

Author: changed

type IngressControllerHTTPHeadersManipulation struct {
	Response []IngressControllerSetOrDeleteHTTPHeader `json:"response,omitempty"`
}
```

```go
// IngressControllerSetOrDeleteHTTPHeader specifies options for set/delete a header.
type IngressControllerSetOrDeleteHTTPHeader struct {
    // name specifies a header name to be set(replace)/append/delete.  Its value must be a valid HTTP header.
    // name as defined in RFC 2616 section 4.2.
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=256
    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
    // +required
    Name string `json:"name"`

    IngressControllerSetOrDeleteHTTPHeaderUnion `json:",inline"`
}
```

```go
// IngressControllerSetOrDeleteHTTPHeaderUnion specifies fields to mention the header details and the action you want to perform on them.
// +union
type IngressControllerSetOrDeleteHTTPHeaderUnion struct {

> 	// action lets you specify headers for Set or Delete.

Reviewer 4: Actually the godoc should use the lowercase "action" to match the json tag (see https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#use-json-field-names-in-godoc).  Otherwise I like the suggestion.  
```suggestion
	// action defines the type of the action to be applied on the header.
```

> 	// action lets you specify headers for Set or Delete.

Reviewer 5: Interesting, didn't know this.
`go doc` (not `godoc`) seems to be fine with the case mismatch btw:
```
$ cat here/f.go 
package here

import (
	"fmt"
)

// here is a function
func Here() {
	fmt.Println("here")
}

$ go doc -all here
package here // import "github.com/alebedev87/dummy/here"

FUNCTIONS

func Here()
    here is a function
```
Not sure which one is used in pkg.go.dev, for instance.

	// +unionDiscriminator
	// +kubebuilder:validation:Enum:=Set;Delete
	// +kubebuilder:validation:Required

> 	Action IngressControllerSetHTTPHeaderAction `json:"action,omitempty"`

Author: eg- Set Action to `Set` or `Delete`

> 	Action IngressControllerSetHTTPHeaderAction `json:"action,omitempty"`

Reviewer 5: Sorry I must have misled you. I meant that the type should not have the specific actions in its name at all. Something like this would do: `IngressControllerHTTPHeaderAction`.

> 	Action IngressControllerSetHTTPHeaderAction `json:"action,omitempty"`

Author: api has been fixed

	// Set defines HTTP headers that should be set.

> 	// If this field is empty, no headers are added.

Reviewer 4: According to https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1027-api-unions/README.md#go-markers, if you specify `// +unionMember` (rather than `// +unionMember,optional`), "then the object will fail validation if the discriminator selects the field but it is nil".  Maybe the validation that the KEP describes isn't actually implemented though.

> 	// If this field is empty, no headers are added.

Author: api has been fixed

	//
	// Note that this option only applies to cleartext HTTP connections
	// and to secure HTTP connections for which the ingress controller
	// terminates encryption (that is, edge-terminated or reencrypt
	// connections).  Headers cannot be set(replace)/append for TLS passthrough
	// connections.
	//
	// +kubebuilder:validation:Optional
	// +optional
	Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`

	// delete defines HTTP headers that should be deleted.

> 	// If this field is empty, no headers are deleted.

Reviewer 4: > // If this field is empty, no header is deleted.

This field cannot be nonempty.

> 	// If this field is empty, no headers are deleted.

Author: api has been updated

	//
	// Note that this option only applies to cleartext HTTP connections
	// and to secure HTTP connections for which the ingress controller
	// terminates encryption (that is, edge-terminated or reencrypt
	// connections).  Headers cannot be set(replace)/append for TLS passthrough
	// connections.
	//
	// +kubebuilder:validation:Optional
	// +optional
	Delete *IngressControllerDeleteHTTPHeader `json:"delete,omitempty"`
}
```

```go
// IngressControllerSetAction defines operation you can perform on headers.
// +kubebuilder:validation:Enum=Set;Delete
type IngressControllerSetHTTPHeaderAction string

const (
    Set    IngressControllerSetHTTPHeaderAction = "Set"
    Delete IngressControllerSetHTTPHeaderAction = "Delete"
)
```

The IngressControllerSetHTTPHeader type has fields for specifying a header value.
```go

> // IngressControllerSetHTTPHeader what value of HTTP header needs to be set.

Author: fixed

type IngressControllerSetHTTPHeader struct {
    // Value specifies a header value.
    // +kubebuilder:validation:Required
    // +required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=4096
    Value string `json:"value"`
}
```

> The IngressControllerDeleteHTTPHeader type is an empty union member .

Reviewer 4: Hm, I guess "empty union member" means that there is no member field.  I have been using the phrase "empty union member" when the field exists but its type is an empty struct.  In that case, "a union member that is an empty struct" would be technically more correct (but it does seem a bit verbose).

> The IngressControllerDeleteHTTPHeader type is an empty union member .

Reviewer 5: Ah ok, now I see what you meant by "empty union member". Yeah, that has a different definition in the conventions doc. I think it's better to avoid using "empty union member" for Delete as it may be misleading for others too. We can rephrase it somehow or remove.

```go
// IngressControllerDeleteHTTPHeader is used specify delete action per header.
type IngressControllerDeleteHTTPHeader struct{
}
```

B) Setting/Replacing/Appending/Deleting HTTP Response headers in the backend section `Plain http backend or backend with TLS terminated at the edge or a 
secure backend with re-encryption.` of the haproxy.config via Route's Spec.

> The Route API's Spec is extended by adding optional `httpSetHeaders` of type `RouteSetHTTPHeaders` :

Author: fixed

```go
type RouteSpec struct {
	...

> 	// httpHeaders defines policy for HTTP headers.

Reviewer: API conventions dictate that the godoc should start with the json tag, this makes sure that generated documentation (eg `oc explain`) is written with the user facing names

	//
	// If this field is empty, the default values are used.
	//
	// +optional

> 	HTTPHeaders *RouteHTTPHeaders `json:"httpHeaders,omitempty" protobuf:"bytes,9,opt,name=httpHeaders"`

Author: it gets automatically generated by the make command for route api

> 	HTTPHeaders *RouteHTTPHeaders `json:"httpHeaders,omitempty" protobuf:"bytes,9,opt,name=httpHeaders"`

Reviewer 5: @Miciah : Do you know why `protobuf` tag is needed for Router API? And should we use protobuf for IngressController API above?

> 	HTTPHeaders *RouteHTTPHeaders `json:"httpHeaders,omitempty" protobuf:"bytes,9,opt,name=httpHeaders"`

Reviewer 4: We generate protobuf for route/v1 but not for config/v1 or operator/v1: https://github.com/openshift/api/blob/54592eea55395af31e82fa9a605eb45523115454/hack/lib/init.sh#L35-L53

However, I'm not sure why that's the case.  @JoelSpeed, do you know why?

> 	HTTPHeaders *RouteHTTPHeaders `json:"httpHeaders,omitempty" protobuf:"bytes,9,opt,name=httpHeaders"`

Reviewer: We generate protobuf for older APIs, those that are served via the `openshift-apiserver` aggregate api server mechanism. For APIs that are served via CRDs, we don't need protobuf definitions so we don't generate them. In general they are considered as deprecated for our uses within OpenShift

}	

```

```go

> // RouteHTTPHeaders describes how to perform action on headers.

Author: api has been changed

type RouteHTTPHeaders struct {
    // headersManipulation specifies options for set/delete headers.
    HeadersManipulation RouteHTTPHeadersManipulation `json:"headersManipulation,omitempty" protobuf:"bytes,1,opt,name=headersManipulation"`
}
```

```go
// RouteHTTPHeadersManipulation defines whether we want to set Response or Request Headers.
// Currently set/delete of HTTP Response Headers is available only.
type RouteHTTPHeadersManipulation struct {
    Response []RouteSetOrDeleteHTTPHeader `json:"response,omitempty" protobuf:"bytes,1,opt,name=response"`
}
```

```go
// RouteSetOrDeleteHTTPHeader specifies options for set/delete a header.
type RouteSetOrDeleteHTTPHeader struct {

>     // name specifies a header name to be set(replace)/append/delete.  Its value must be a valid HTTP header.

Reviewer 4: As I mentioned in another comment, using the json field name is appropriate (see https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#use-json-field-names-in-godoc).

}
```

```go
// RouteSetHTTPHeaderAction defines operation you can perform on headers.
// +kubebuilder:validation:Enum=Set;Delete
type RouteSetHTTPHeaderAction string

const (
    Set    RouteSetHTTPHeaderAction = "Set"
    Delete RouteSetHTTPHeaderAction = "Delete"
)
```

The RouteSetHTTPHeader type has fields for specifying a header name, header value and action.
```go
// RouteSetHTTPHeader specifies what value of HTTP header needs to be set.
type RouteSetHTTPHeader struct {
    // Value specifies a header value.
    // +kubebuilder:validation:Required
    // +required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=4096
    Value string `json:"value" protobuf:"bytes,2,opt,name=value"`
}
```

```go
// RouteHTTPHeaderAction is a policy for setting, replacing or appending HTTP headers.
//
// +kubebuilder:validation:Enum=Append;Replace
type RouteHTTPHeaderAction string

const (
	// AppendHTTPHeaderPolicy appends the header, preserving any existing header.
	AppendHTTPHeader RouteHTTPHeaderPolicy = "Append"
	// ReplaceHTTPHeaderPolicy sets the header, removing any existing header.
	ReplaceHTTPHeaderPolicy RouteHTTPHeaderPolicy = "Replace"
)

> ```

Author: updated

> The RouteDeleteHTTPHeader is an empty union member .

Author: updated

```go
// RouteDeleteHTTPHeader is used specify delete action per header.
type RouteDeleteHTTPHeader struct {
}
```

### Implementation Details/Notes/Constraints
A) Setting HTTP headers by cluster admin via IngressController's Spec.

Implementing this enhancement requires changes in the following repositories:

* openshift/api
* openshift/cluster-ingress-operator
* openshift/router

The router configures HAProxy using a configuration template.  The template uses
environment variables as input parameters.  The enhancement adds a new
environment variables, `ROUTER_SET_HTTP_HEADERS`, which specify the headers to set in the

> frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config.

Author: done

The ingress operator sets this variable based on the IngressController's
`spec.httpSetHeaders` field value.

> The values of `ROUTER_SET_HTTP_HEADERS` are comma-delimited lists of percent encoded values of
> the form `name:value` separated by comma.

Reviewer 5: ```suggestion
The value of `ROUTER_SET_HTTP_HEADERS` environment variable is a comma-delimited list of percent encoded pairs of the form `name:value`.
```

For example, if `ROUTER_SET_HTTP_HEADERS` has percent encoded value `X-Frame-Options%3A+DENY,X-XSS-Protection%3A+1%3Bmode%3Dblock` 
of `X-Frame-Options: DENY:Set,X-XSS-Protection: 1;mode=block` then the template adds the following stanzas to
the frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config:
```conf
    http-response set-header X-Frame-Options: DENY
    http-response set-header X-XSS-Protection: 1;mode=block    
```

This feature's implementation will be similar to an existing feature called logging headers which was implemented in 
[router](https://github.com/openshift/router/commit/40ee0ba472a4ac503f1e6c97d4464f559101c97c) and 
[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/commit/c5139cb07c3a99a8a3b18a6c9321310add51b075).

B) Deleting HTTP headers by cluster admin via IngressController's Spec.

Implementing this enhancement requires changes in the following repositories:

* openshift/api
* openshift/cluster-ingress-operator
* openshift/router

The router configures HAProxy using a configuration template.  The template uses
environment variables as input parameters.  The enhancement adds a new

> environment variables, `ROUTER_DELETE_HTTP_HEADERS`, which specify the headers to set in the

Reviewer 4: I think you want to use the same environment variable for `spec.httpHeaders.headersManipulation.response[*].set` and `spec.httpHeaders.headersManipulation.response[*].delete` in order to preserve ordering.

Author: i think two env vars is fine. If we want only one then I will have to append Action to name:value:action in case of Set and name:action incase of Delete

Author: i think I get your point but then we will have pass Action as mentioned above in the env var.

frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config.
The ingress operator sets this variable based on the IngressController's
`spec.httpDeleteHeaders` field value.

> The values of `ROUTER_DELETE_HTTP_HEADERS` are comma-delimited lists of values of
> the form `name` separated by comma.

Reviewer 5: ```suggestion
The values of `ROUTER_DELETE_HTTP_HEADERS` are comma-delimited lists of values of
the form `name` separated by comma.
```
```suggestion
The value of `ROUTER_DELETE_HTTP_HEADERS` environment variable is a comma-delimited list of percent encoded values of the form `name`.
```

The template translates these values into an appropriate `http-response del-header` stanzas.
For example, if `ROUTER_DELETE_HTTP_HEADERS` has a value `X-Frame-Options,X-XSS-Protection`
then the template adds the following stanzas to the frontend fe_no_sni, frontend public, frontend fe_sni  section
of the haproxy.config:
```conf
    http-response del-header X-Frame-Options
    http-response del-header X-XSS-Protection   
```

C) Setting HTTP headers via Route's Spec by Users.

Implementing this enhancement requires changes in the following repositories:

* openshift/api
* openshift/router
* openshift/openshift-apiserver

The router configures HAProxy using a configuration template. The template is populated with values in ServiceAliasConfig .

> The enhancement adds a field in ServiceAliasConfig called `SetReplaceAppendHTTPResponseHeaders`, which specify the response headers to set in the

Reviewer 5: SetReplaceAppend? We only do `Set` now.

> backend section `Plain http backend or backend with TLS terminated at the edge or a secure backend with re-encryption.` of the
> haproxy.config at this [code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609).

Reviewer 5: ```suggestion
the haproxy.config sections corresponding to the following backends: plain http backend, backend with TLS terminated at the edge, secure backend with re-encryption ( [code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609)).
```

The router sets this variable based on the Route's

> `spec.httpSetHeaders` field value. Service AliasConfig will be populated at this [code location](https://github.com/openshift/router/blob/master/pkg/router/template/router.go#L927)

Reviewer 5: ```suggestion
`spec.httpHeaders.headersManipulation.reponse[*].set` field values. Service AliasConfig will be populated at this [code location](https://github.com/openshift/router/blob/master/pkg/router/template/router.go#L927)
```

```go
// ServiceAliasConfig is a route for a service.  Uniquely identified by host + path.
type ServiceAliasConfig struct {
    ...
	// For route based setting of headers

> 	SetReplaceAppendHTTPResponseHeaders []SetReplaceAppendHTTPHeader

Reviewer 5: Replace Append, again.

Author: fixed

	DeleteHTTPResponseHeaders           []DeleteHTTPHeader
}
```
```go
// SetReplaceAppendHTTPHeader specifies an HTTP header that should be set/replace/append
type SetReplaceAppendHTTPHeader struct {
	// Name specifies an HTTP header name.
	Name string

	// Value specifies the header value.
	Value string

	// Action specifies the action to be performed.
	Action string
}

type DeleteHTTPHeader struct {
	// Name specifies an HTTP header name.
	Name string
}
```

D) Deleting HTTP headers via Route's Spec by Users.

Implementing this enhancement requires changes in the following repositories:

* openshift/api
* openshift/router
* openshift/openshift-apiserver

The router configures HAProxy using a configuration template. The template is populated with values in ServiceAliasConfig .

> The enhancement adds a field in ServiceAliasConfig called `DeleteHTTPResponseHeaders`, which specify the response headers to be deleted in the
> backend section `Plain http backend or backend with TLS terminated at the edge or a secure backend with re-encryption.` of the
> haproxy.config at this [code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609).

Reviewer 5: ```suggestion
The enhancement adds a field in ServiceAliasConfig called `DeleteHTTPResponseHeaders`, which specifies the response headers to be deleted in the haproxy.config secrtions corresponding to the following backends: plain http backend, backend with TLS terminated at the edge, secure backend with re-encryption ([code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609)).
```

Author: done

The router sets this variable based on the Route's

> `spec.httpDeleteHeaders` field value. ServiceAliasConfig will be populated at this [code location](https://github.com/openshift/router/blob/master/pkg/router/template/router.go#L927)

Reviewer 5: ```suggestion
`spec.httpHeaders.headersManipulation.response[*].delete` field values. ServiceAliasConfig will be populated at this [code location](https://github.com/openshift/router/blob/master/pkg/router/template/router.go#L927)
```

Author: done

  Answer: Yes, this enhancement won't allow the user to set/replace, append or delete HSTS headers. We already a way to set this which is via route 
  annotation. Providing one more way to add this can cause issues.

- Would we be able to [Replace a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-a-header-by-using-a-regular-expression)?
  Answer: No, this is out of scope for this enhancement.

- Would we be able to [Replace part of a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-part-of-a-header-by-using-a-regular-expression) ?
  Answer:  No, this is out of scope for this enhancement.

### Test Plan

The controller that manages the IngressController Deployment and related
resources has unit test coverage; for this enhancement, the unit tests are
expanded to cover the additional functionality.

The operator has end-to-end tests; for this enhancement, a test is added that
(1) creates an IngressController that specifies certain response

> headers should be set(replace) or append or deleted., (2) sends an HTTP request to one of the standard

Reviewer 5: Can you please do a search for `replace` and `append` words in this EP and make sure they are not mentioned anywhere?

Author: done

routes (such as the console route), and (3) verifies that the expected headers

> are logged.

Reviewer 5: Expected headers are logged? What does this mean exactly? How exactly we would check for the presence of the headers we set or the absence of the headers we delete?

Author: curl -vk <console route> shall show the headers set in response.

  name: hello-openshift-edge
  namespace: hello-openshift
spec:
  subdomain: hello-openshift 
  tls:
    termination: edge
  to:
    kind: Service
    name: hello-openshift
  httpHeaders:
    headersManipulation:
      action: Delete
      delete:
        response:
        - name: X-Frame-Options
        - name: X-XSS-Protection
```
#### Variation [optional]

### API Extensions

A) Set(Replace)/Append/Delete HTTP headers in the global section of haproxy.config via IngressController's Spec.
The existing Ingress Controller API `HTTPHeaders *IngressControllerHTTPHeaders` is extended by adding an field for action to be performed,
and field for headers to be set/replaced/appended/deleted.

```go
// IngressControllerSpec is the specification of the desired behavior of the
// IngressController.
type IngressControllerSpec struct {
    // httpHeaders defines policy for HTTP headers.
    //
    // If this field is empty, the default values are used.
    //
    // +optional
    HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`
}
```

```go

type IngressControllerHTTPHeaders struct {
    // headersManipulation specifies options for set/replace/append/delete headers.

>     // +union

Reviewer 5: `+union` has to be set on the definition of `IngressControllerHeadersManipulation` below, not here.

Reviewer 5: Done.

Author: api changes are updated

    // delete defines HTTP headers that should be deleted.
    // If this field is empty, no headers are deleted.
    //
    // Note that this option only applies to cleartext HTTP connections
    // and to secure HTTP connections for which the ingress controller
    // terminates encryption (that is, edge-terminated or reencrypt
    // connections).  Headers cannot be set(replace)/append for TLS passthrough
    // connections.
    //
    // +kubebuilder:validation:Optional
    // +optional
    Delete *IngressControllerDeleteHTTPHeaders `json:"delete,omitempty"`
}
```

The IngressControllerSetHTTPHeaders type has a Response field,
of type []IngressControllerSetHTTPHeader, for specifying headers to set/append:

```go
// The IngressControllerSetHTTPHeaders type has a Response field,
// type []IngressControllerSetHTTPHeader, for specifying headers to set(replace) or append:

type IngressControllerSetHTTPHeaders struct {
    // response specifies which HTTP response headers to set(replace) or append.
    //
    // If this field is empty, no response headers are set(replace) or append.
    //
    // +nullable
    // +optional

>     Response []IngressControllerSetHTTPHeader `json:"response,omitempty"`

Reviewer 5: `Response` is repeated in Set and Delete structures. It can be moved up to `headerManipulation` to reduce the redundancy. Also, it seems to be more fundamental as parameter than set and delete.

Reviewer 5: Done. The naming: `headerManipulation` tag can still be discussed though, maybe a better name will come up into mind.

Author: api changes are updated

```go
type IngressControllerSetHTTPHeader struct{
    // name specifies a header name.  Its value must be a valid HTTP header 
    // name as defined in RFC 2616 section 4.2.
    //	
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=256
    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$" 
    // +required 
    Name string `json:"name"`

    // Value specifies a header value.
    // +kubebuilder:validation:Required 
    // +required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=4096
    Value string `json:"value"`

    // Action specifies a header value.
    // +kubebuilder:validation:Required
    // +required
    // +kubebuilder:validation:Enum=Append;Replace
    Action IngressControllerHTTPHeaderAction `json:"action"`
}
```

This defines Actions to be perfomed on headers when action field is `Set` in the IngressControllerHTTPHeaders. 
```go
// IngressControllerHTTPHeaderAction is a policy for setting HTTP headers.
// +kubebuilder:validation:Enum=Append;Replace

> type IngressControllerHTTPHeaderAction string

Reviewer 5: This is not accurate anymore, you use `Set` and `Delete` now.

Author: For setting headers  we either Append or Replace(Set). So we need this.

Reviewer 5: Sorry confused it with a similar type which for some reason is not (but should) present in the EP: `IngressControllerHTTPHeaderAction`.

Author: api changes are updated

   deleted at the global level when forwarding connections to all applications from clients.
3. Enable the route owner to specify HTTP response headers in the Route spec (when it is not a passthrough Route) for their application
   that should be set or deleted when forwarding a response to the client.
4. Enable the route owner to specify HTTP request headers in the Route spec (when it is not a passthrough Route) for their application
   that should be set or deleted when forwarding a connection to the application from a client.
5. Define an immutable/hard-coded deny list of headers which are not allowed to be set or deleted.
   Eg- we should prohibit specifying strict-transport-security using the new API, for the following reasons:
   HSTS is already configurable through a route annotation.
   We already have an API for restricting HSTS in the cluster config.
   Adding a second way to configure HSTS would significantly complicate this enhancement.
6. Setting dynamic header values for http response. Allowed sample fetchers are res.hdr, ssl_c_der. Allowed converters are lower, base64.

> 7. Setting dynamic header values for http request. Allowed sample fetchers are req.hdr, ssl_c_der. Allowed converters are lower, base64.

Reviewer 3: Goals 6 and 7 were added after the RFE was accepted and tend to complicate this enhancement, as well as greatly increase time spent on code review, documentation, and QE.  Can we leave this as an iterative update for another time?

Author: This has been already considered now as some customer wanted it.

```

#### Variation [optional]

### API Extensions

A) Set/Delete HTTP response headers in the frontend sections of haproxy.config via IngressController's Spec.
The existing Ingress Controller API `HTTPHeaders *IngressControllerHTTPHeaders` is extended by adding a field for the header manipulations.

```go
// IngressControllerSpec is the specification of the desired behavior of the
// IngressController.
type IngressControllerSpec struct {
    // httpHeaders defines policy for HTTP headers.
    //
    // If this field is empty, the default values are used.
    //
    // +optional
    HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`
}
```

```go
// IngressControllerHTTPHeaders specifies how the IngressController handles
// certain HTTP headers.
type IngressControllerHTTPHeaders struct {
    // actions specifies options for performing action on headers.
    // Note that this option only applies to cleartext HTTP connections
    // and to secure HTTP connections for which the ingress controller
    // terminates encryption (that is, edge-terminated or reencrypt
    // connections).  Headers cannot be set or deleted for TLS passthrough
    // connections.

>     // Setting HSTS header is supported to be set via route annotation haproxy.router.openshift.io/hsts_header and not via this API.

Reviewer 3: I left a comment in the API.  Setting HSTS is a whole system in the IngressSpec.

Author: OK

Author: added here.

---
title: set-replace-append-delete-deny-headers
authors:
  - "@miheer"
reviewers:
  - "@frobware"
  - "@gcs278"
  - "@Miciah"
approvers:
  - "@frobware"
  - "@gcs278"
  - "@Miciah"
api-approvers:
  - "@JoelSpeed"
  - "@deads2k"
  - "@frobware"
  - "@gcs278"
  - "@Miciah"
creation-date: 2022-12-02
last-updated: 2022-12-14
tracking-link:
  - "https://issues.redhat.com/browse/NE-982"
  - "https://issues.redhat.com/browse/RFE-464"
status: TODO
see-also:
  - ""
replaces:
superseded-by:
---

# Set, Delete, Deny HTTP Response headers

## Summary

This enhancement extends the IngressController API and Route API to allow the user to set or delete
HTTP request/response headers. Using the new API, users may specify a list of

> HTTP request/response headers either via IngressController API or Route API that should be set or deleted.

Author: done

> Also, this enhancement creates an immutable(hard-coded) list of HTTP response headers which can't be set or deleted.

Author: done

## Motivation

> Certain HTTP request/response headers may need to be set, deleted for compliance reasons or other reasons. In OpenShift 3, cluster

Author: done

administrators could use a custom HAProxy configuration template to set or delete headers, but using a custom
template is not permitted in OpenShift 4.
Certain HTTP response headers like HSTS (Strict-Transport-Security) need to be denied as we already provide a way to set them
via route annotation.

### User Stories

1. As a cluster administrator, I need to configure an IngressController to set/delete certain HTTP request/response headers in order to

>    satisfy compliance requirements at global level to send/not to send them to the client which has called any route from IngressController's domain.

Author: done

   Example 1: As a cluster administrator I want to set http response headers like `X-XSS-Protection "1; mode=block"` , `X-Content-Type-Options "nosniff"` and `X-Frame-Options "SAMEORIGIN"` to be sent to all routes.

>    Example 2: As a cluster administrator I want to set http request headers like `Content-Location "/my-first-blog-post"` , `X-SSL-Client-Cert "%{+Q}[ssl_c_der,base64]"` and `X-Target "%[req.hdr(host),lower]"` to be sent to all applications.

Author: done

>    Example 3: As a cluster administrator I want to delete http response headers like `X-XSS-Protection` to be not sent to the client.

Author: done

>    Example 4: As a cluster administrator I want to delete http request headers like `Content-Location` to be not sent to the application.

Author: done

2. As a user, I need to configure Route to set/delete HTTP request/response headers to send/not send them to the client which has called the route.
   Example 1: As a user I want to set http response headers like `X-XSS-Protection "1; mode=block"` , `X-Content-Type-Options "nosniff"` and `X-Frame-Options "SAMEORIGIN"` to be sent to the client.
   Example 2: As a user I want to set http request headers like `Content-Location "/my-first-blog-post"` , `X-SSL-Client-Cert "%{+Q}[ssl_c_der,base64]"` and `X-Target "%[req.hdr(host),lower]" to be sent to the application.`
   Example 3: As a user I want to delete http response headers like `X-XSS-Protection` to be not sent to the client.

>    Example 4: As a user I want to delete http request headers like `Content-Location to be not sent to the application.`

Author: done

   deleted at the global level when forwarding connections to all applications from clients.
3. Enable the route owner to specify HTTP response headers in the Route spec (when it is not a passthrough Route) for their application
   that should be set or deleted when forwarding a response to the client.
4. Enable the route owner to specify HTTP request headers in the Route spec (when it is not a passthrough Route) for their application
   that should be set or deleted when forwarding a connection to the application from a client.
5. Define an immutable/hard-coded deny list of headers which are not allowed to be set or deleted.
   Eg- we should prohibit specifying strict-transport-security using the new API, for the following reasons:
   HSTS is already configurable through a route annotation.
   We already have an API for restricting HSTS in the cluster config.
   Adding a second way to configure HSTS would significantly complicate this enhancement.
6. Setting dynamic header values for http response. Allowed sample fetchers are res.hdr, ssl_c_der. Allowed converters are lower, base64.

> 7. Setting dynamic header values for http request. Allowed sample fetchers are req.hdr, ssl_c_der. Allowed converters are lower, base64.

Author: done

> 8. The header set using route will over-ride the header set in ingress controller if the headers defined both were same.

Author: done

### Non-Goals

1. Validation that each specified value is valid for the respective header (e.g. require that either "DENY" or "SAMEORIGIN" is 
   specified for [the `x-frame-options` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options)).
2. Setting load balancing algorithm using header values. The HAproxy configuration provides a [load balancing algorithm](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-balance) which relies on the
   check against the incoming request's header values. Even though this enhancement adds the header manipulation capability it
   doesn't aim at providing (or enhancing) the load balancing API for the IngressController or Route.  Therefore such a 
   load balancing algorithm is out of scope for this enhancement and should be handled with a separate RFE.

> 3. Configuring buffer sizes by referring headers. Eg- tune.bufsize and tune.maxrewrite

Author: done

4. Modifying headers for a passthrough route. 
5. Providing dedicated API knobs for the [appending](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response%20add-header) or 
   [replacing](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response%20replace-header) headers. 
   The headers will be modified according to the [set-header HAProxy configuration stanza](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response%20set-header): add header 
   if it doesn't exist and remove before adding if it does.

> 6. Allowing sample fetchers other than res.hdr, ssl_c_der and converters other than lower, base64 in the dynamic header value.

Author: done

> 7. To not allow the route owner to over-ride the header set in ingress controller if the headers defined in the route and ingress controller were same.

Author: done

> 7. To not allow the route owner to over-ride the header set in ingress controller if the headers defined in the route and ingress controller were same.

Reviewer 4: What happened to this non-goal ("Providing a way to prohibit the Route owner from overriding the configuration for an IngressController if the Route and IngressController both set or delete a header with the same name")?

> 7. To not allow the route owner to over-ride the header set in ingress controller if the headers defined in the route and ingress controller were same.

Author: not sure how this was removed

> 7. To not allow the route owner to over-ride the header set in ingress controller if the headers defined in the route and ingress controller were same.

Reviewer 4: https://github.com/openshift/enhancements/compare/ecce6af5acd88f369db08634d43322c62073dc40..9046f62be67a7e43a35491c60d1f8e95fa1be235 re-added this goal, so my previous comment has been addressed.  Thanks!

                set:
                  value: DENY
            - name: X-Cache-Info
              action:
                type: Set
                set:
                  value: "not cacheable; meta data too large"
            - name: X-XSS-Protection
              action:
                type: Delete
            - name: X-Source
              action:
                type: Set
                set:
                  value: "%[res.hdr(X-Value),lower]"
            request:
            - name: Content-Location
              action:
                type: Set
                set:
                  value: /my-first-blog-post
            - name: X-SSL-Client-Cert
              action:
                type: Set
                set:
                  value: "%{+Q}[ssl_c_der,base64]"
            - name:   Content-Language
              action:
                type: Delete
            - name: X-Target
              action:
                type: Set
                set:
                  value: "%[req.hdr(host),lower]"
```

#### Variation [optional]

### API Extensions

A) Set/Delete HTTP response headers in the frontend sections of haproxy.config via IngressController's Spec.

> The existing Ingress Controller API `HTTPHeaders *IngressControllerHTTPHeaders` is extended by adding a field for the header manipulations.

Author: done

```go
// IngressControllerHTTPHeaderActionUnion specifies an action to take on an HTTP header.
// +union
type IngressControllerHTTPHeaderActionUnion struct {
    // type defines the type of the action to be applied on the header.
    // Possible values are Set and Delete.
    // Set allows you to set http request and response headers.
    // Delete allows you to delete http request and response headers
    // +unionDiscriminator
    // +kubebuilder:validation:Enum:=Set;Delete
    // +kubebuilder:validation:Required
    Type IngressControllerHTTPHeaderActionType `json:"type"`

    // set specifies how the HTTP header should be set.
    // If this field is empty, no action is taken.
    // +optional
    // +unionMember
    Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`
}
```

```go
// IngressControllerHTTPHeaderActionType defines actions that can be performed on HTTP headers.
type IngressControllerHTTPHeaderActionType string

const (
    // Set specifies that an HTTP header should be set.
    Set IngressControllerHTTPHeaderActionType = "Set"
    // Delete specifies that an HTTP header should be deleted.
    Delete IngressControllerHTTPHeaderActionType = "Delete"
)
```

> The IngressControllerSetHTTPHeader type has fields for specifying a header value.

Author: done

```go
// IngressControllerSetHTTPHeader defines the value which needs to be set on an HTTP header.
type IngressControllerSetHTTPHeader struct {
    // value specifies a header value.
    // Dynamic values can be added. The value will be interpreted as an HAProxy format string as defined in
    // http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#8.2.6 and may use HAProxy's %[] syntax and
    // otherwise must be a valid HTTP header value as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2
    // Sample fetchers allowed are "req.hdr" and "ssl_c_der".
    // Converters allowed are "lower" and "base64".
    // Examples of header value - "%[req.hdr(host),lower]", "%[req.hdr(host),base64]", "{+Q}[ssl_c_der,base64]", "{+Q}[ssl_c_der,lower]"
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=4096
    Value string `json:"value"`
}
```

B) Set/Delete HTTP response headers in the backend section (plain http backend or backend with TLS terminated at the edge or a
secure backend with re-encryption) of the haproxy.config via Route's Spec.

> The Route API's Spec is extended by adding optional `HTTPHeaders` field of type `RouteHTTPHeaders` :

Author: done

// +union
type RouteHTTPHeaderActionUnion struct {
    // type defines the type of the action to be applied on the header.
    // Possible values are Set and Delete.
    // Set allows you to set http request and response headers.
    // Delete allows you to delete http request and response headers
    // +unionDiscriminator
    // +kubebuilder:validation:Enum:=Set;Delete
    // +kubebuilder:validation:Required
    Type RouteHTTPHeaderActionType `json:"type" protobuf:"bytes,11,opt,name=type,casttype=RouteHTTPHeaderActionType"`

    // set defines the HTTP header that should be set: added if doesn't exist or replaced if does.
    // If this field is empty, no header is added or replaced.
    // +optional
    // +unionMember
    Set *RouteSetHTTPHeader `json:"set,omitempty" protobuf:"bytes,9,opt,name=set"`
}
```

```go
// RouteHTTPHeaderActionType defines actions that can be performed on HTTP headers.
type RouteHTTPHeaderActionType string

const (
    // Set specifies that an HTTP header should be set.
    Set RouteHTTPHeaderActionType = "Set"
    // Delete specifies that an HTTP header should be deleted.
    Delete RouteHTTPHeaderActionType = "Delete"
)
```

> The RouteSetHTTPHeader type has fields for specifying a header name, header value and action.

Author: done

    // Sample fetchers allowed are "req.hdr" and "ssl_c_der".
    // Converters allowed are "lower" and "base64".
    // Examples of header value - "%[req.hdr(host),lower]", "%[req.hdr(host),base64]", "{+Q}[ssl_c_der,base64]", "{+Q}[ssl_c_der,lower]"
    // Note: Any change to regex mentioned below shall be reflected in the crd validation of route in library-go or vice-versa.
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=2048
    Value string `json:"value" protobuf:"bytes,2,opt,name=value"`
}
```

### Implementation Details/Notes/Constraints
A) Setting HTTP headers by cluster admin via IngressController's Spec.

Implementing this enhancement requires changes in the following repositories:

* openshift/api
* openshift/cluster-ingress-operator
* openshift/router

The router configures HAProxy using a configuration template.  The template uses
environment variables as input parameters.  The enhancement adds two new
environment variables, `ROUTER_HTTP_RESPONSE_HEADERS`, `ROUTER_HTTP_REQUEST_HEADERS` which specifies the request and response http headers to set or deleted in the

> following frontend sections of haproxy.config: `public`, `fe_no_sni`,`fe_sni`.

Author: done

The value of `ROUTER_HTTP_RESPONSE_HEADERS` or `ROUTER_HTTP_REQUEST_HEADERS` is a comma-delimited list of percent encoded  triplets of the form `name:value:action` for action `Set` and `name:action` pairs for action `Delete`.
The template translates these values into the appropriate `http-response set-header` stanzas.
For example, if `ROUTER_HTTP_RESPONSE_HEADERS` or `ROUTER_HTTP_REQUEST_HEADERS` has base64 encoded value `"WC1GcmFtZS1PcHRpb25z:REVOWQ==:U2V0,WC1YU1MtUHJvdGVjdGlvbg==:MTttb2RlPWJsb2Nr:U2V0,eC1mb3J3YXJkZWQtY2xpZW50LWNlcnQ=:JXsrUX1bc3NsX2NfZGVyLGJhc2U2NF0=:U2V0,WC1GcmFtZS1PcHRpb25z:RGVsZXRl,WC1YU1MtUHJvdGVjdGlvbg==:RGVsZXRl"`
of `X-Frame-Options:DENY:Set,X-XSS-Protection:1;mode=block:Set,x-forwarded-client-cert:%{+Q}[ssl_c_der,base64]:Set,X-Frame-Options:Delete,X-XSS-Protection:Delete` and `QWNjZXB0:dGV4dC9wbGFpbiwgdGV4dC9odG1s:U2V0,QWNjZXB0LUVuY29kaW5n:RGVsZXRl` of `Accept:text/plain, text/html:Set,Accept-Encoding:Delete` respectively then the template adds the following stanzas to

> the frontend fe_no_sni, frontend public, frontend fe_sni  sections of the haproxy.config:

Author: added percent encoding as it looked more clean and readable.

```conf
    http-response set-header X-Frame-Options 'DENY'
    http-response set-header X-XSS-Protection '1;mode=block'
    http-response set-header x-forwarded-client-cert '%{+Q}[ssl_c_der,base64]'
    http-response del-header X-Frame-Options
    http-response del-header X-XSS-Protection
    http-request set-header Accept 'text/plain, text/html'
    http-request del-header Accept-Encoding
```

> Header values are sanitized so that haproxy does not throw an error while parsing the haproxy template for http-response and http-request stanzas.

Author: done

Any single quotes provided within the header value will be escaped and replaced with '\'' and then the value is encapsulated with single quotes ''.
Eg: Header value `"abc"\ 'def'` is sanitized to `'"abc"\ '\''def'\'''`

>     Header value `%[req.hdr(Host)] if foo` is sanitized to '%[req.hdr(Host)] if foo'

Author: done

The following shows how it will get set in haproxy.config
```conf
http-response set-header X-Frame-Options '\"abc\"\ "def"'
http-response set-header X-Server '%[req.hdr(Host)] if foo'
```

> Use of a single environment variable `ROUTER_HTTP_RESPONSE_HEADERS` for http response header  and `ROUTER_HTTP_REQUEST_HEADERS`for http request header respectively for both action `Set` and `Delete` will help preserve ordering.

Author: done

This feature's implementation will be similar to an existing feature called logging headers which was implemented in 
[router](https://github.com/openshift/router/commit/40ee0ba472a4ac503f1e6c97d4464f559101c97c) and 
[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/commit/c5139cb07c3a99a8a3b18a6c9321310add51b075).

B) Setting or Deleting HTTP headers via Route's Spec by Users.

Implementing this enhancement requires changes in the following repositories:

* openshift/api
* openshift/router
* openshift/openshift-apiserver

> The router configures HAProxy using a configuration template. The template is populated with values from ServiceAliasConfig .

Author: done

> The enhancement adds a field in ServiceAliasConfig called `SetHTTPResponseHeaders` or `DeleteHTTPResponseHeaders` or `SetHTTPRequestHeaders` or `DeleteHTTPRequestHeaders` based on the action `Set` or `Delete` respectively, which specifies the request or response headers to be set or deleted in the

Author: done

> the haproxy.config sections corresponding to the following backends: plain http backend, backend with TLS terminated at the edge, secure backend with re-encryption ( [code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609)).

Reviewer 4: Still need to fix this to use a permalink: https://github.com/openshift/enhancements/blob/ae5a9e6428596afcad8ad1c83f801768480c308b/enhancements/ingress/set-delete-http-headers.md#L601

	SetHTTPRequestHeaders              []SetHTTPHeader
	DeleteHTTPRequestHeaders           []DeleteHTTPHeader
}
```
```go
// SetHTTPHeader specifies an HTTP header that should be set.
type SetHTTPHeader struct {
	// Name specifies an HTTP header name.
	Name string

	// Value specifies the header value.
	Value string

	// Action specifies the action to be performed.
	Action string
}

type DeleteHTTPHeader struct {
	// Name specifies an HTTP header name.
	Name string
}
```

In this case the ordering of actions will be maintained as we will be having access to the RouteSpec inside router. So, we will run a loop
on either `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type` or `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type` which will have the desired
sequence.

C) library-go code will be having validations for checking route api fields.
We will be creating an immutable/harcoded deny list of HTTP Response headers in the [library-go](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141)

> which won't be allowed to be set/delete. Eg - HSTS.

Author: done

> Only 64 items in request and response header list will be allowed.

Author: The api requires you to add a limit when you run the make command. So had to add this. It is kind of implementation detail.

Minimum length of the header value would have to be greater than 0 and less than 4096.
Sample fetchers for response dynamic header value allowed are "res.hdr" and "ssl_c_der".
Sample fetchers for request dynamic header value allowed are "req.hdr" and "ssl_c_der".
Converters allowed for request and response are "lower" and "base64".

> Actions like Set/Delete for headers can't be set for passthrough routes.

Author: done

D) route/v1 of openshift/api will contain the validations using kubebuilder tags and CEL.
The validations will be same as mentioned in point C).

E) operator/v1/types_ingress.go of openshift/api will contain the validations using kubebuilder tags and CEL.
The validations will be same as mentioned in point C) except the one for passthrough routes.

F) To validate the header value it has to match a regex.

For ingress operator CR and route CR(microshift) this is defined on the request and response list of header in openshift/api.

> For route CR(OCP) this is defined in library-go.

Author: done

This will be the regex:

> ```sh

Author: we can add it in go as this is used in go lang

request header:
`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:req\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```
```sh
response header:
`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```

Allowed dynamic values for request would be -
```conf
%[req.hdr(host),lower]
%[req.hdr(host),base64]
%[req.hdr(Host)] if foo
```
Allowed dynamic values for response would be -
```conf
%[res.hdr(X-Acl-Key),lower]
%[res.hdr(X-Acl-Key),base64]
%[req.hdr(Host)] if foo
```
Dynamic values allowed for both request and response would be -
```conf
%{+Q}[ssl_c_der,base64]
```

For regular header which are not dynamic one can pass values as mentioned in unit test `TestHeaderValues` with `validInput:true`
present in `pkg/route/validation/validation_test.go`

Please note that all values will be sanitized before passing them to the haproxy.conf where a single quote present in value is escaped and the whole value is single quoted which will help
haproxy not to crash in the event of an invalid input. This sanitize logic will be present in router.

> Quote using single quotes, we will do the following:

Author: done

Replace ' with '\''.
Wrap the entire string in single quotes.
eg: "abc"\\ 'def'  shall it be converted to '"abc"\\ '\''def'\'''

### Risks and Mitigations

### Drawbacks

## Design Details

### Open Questions [optional]

> - Would this enhancement cover setting or deleting request headers ?
>   Answer: Yes we would be covering setting or deleting of request headers.

Reviewer 4: This still needs some cleanup and expansion.  It's been a while, but I think this suggested edit addresses what the question was originally about:
````suggestion
- Will this enhancement cover setting or deleting request headers in addition to response headers?
  Answer: Yes, we will be covering setting or deleting of request headers.
  This is required for the following use-case.  As a cluster admin or route owner, I want to enable
  mutual TLS but replace the "x-ssl-client-der" request header that OpenShift router adds with an
  "x-forwarded-client-cert" request header that my application checks for, so that I can migrate my
  applications that rely on mutual TLS and use the latter header name.  The following example
  replaces "x-ssl-client-der" with "x-forwarded-client-cert":
  ```yaml
  spec:
    httpHeaders:
      headersManipulation:
        request: 
        - name: x-ssl-client-der
          action: Delete
        - name: x-forwarded-client-cert
          action: Set
          set:
            value: "%{+Q}[ssl_c_der,base64]"
  ```
````

Author: done

> - Would we be able to set load balancing algorithm based on headers ?
>   Answer: No, this out of scope for this enhancement so, it will be considered as a non-goal. Please check point number 5 in the non-goals section for more details.

Reviewer 4: It looks like the numbers changed, and you mean to point to what is now number 2 ("Setting load balancing algorithm using header values.").  It seems odd to me that you put the detailed information in the non-goals section and then added a reference here, but the explanation in the non-goals section is clear enough.  
```suggestion
- Will this enhancement cover setting the load-balancing algorithm based on headers?
  Answer: No, this out of scope for this enhancement, so it will be considered as a non-goal. Please check point number 2 in the non-goals section for more details.
```

Author: done

> - Would we be able to delete or append headers  ?
>   Answer: Yes, this enhancement will cover deletion and not appending of HTTP Response Headers as per the
>   [epic](https://issues.redhat.com/browse/RFE-464). 
>   The reason is "add-header" appends an HTTP header field whose name is specified in and whose value is defined by 
>   which follows the log-format rules. This may be used to send a cookie to a
>   client for example, or to pass some internal information. This rule is not final, so it is possible to add other 
>   similar rules. Note that header addition is performed immediately, so one rule might reuse the resulting header 
>   from a previous rule.

Reviewer 4: You still haven't explained why appending isn't covered, so I'll make a guess. 
```suggestion
- Will this enhancement cover deleting or appending headers?
  Answer: Yes and no.  This enhancement will cover deletion but not appending of
  HTTP response headers as per the requirements from
  [RFE-464](https://issues.redhat.com/browse/RFE-464).  While this RFE mentions
  use-cases that require the ability to delete headers, there is no use-case
  from this RFE that requires the ability to append headers.  Once such a
  use-case arises, we may add the option to append headers, with due
  consideration of any additional requirements deriving from the use-case
  itself or considerations around the handling of comma-separated values.  (Note
  that a use-case involving the `x-forwarded-*` headers, which is not mentioned
  on the RFE, is described below.)
```

Author: This is mentioned in the EP -
  For the headers like `X-Forwarded-For` which might need values to be appended we have a different API for that.
  This API is specifically meant just either set or delete an HTTP Request and Response header.
  In this API if someone sets an HTTP header with a value we expect them to set a complete value in advance rather than appending it later to the header.

  "set-header" does the same as "add-header" except that the header name is first removed if it existed. This is 
  useful when passing security information to the server, where the header must not be manipulated by external users.
  So, that is the reason we don't need add-header.
  References - https://wohugb.gitbooks.io/haproxy/content/config/http-response.html
       https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/
  For the headers like `X-Forwarded-For` which might need values to be appended we have a different API for that.
  This API is specifically meant just either set or delete an HTTP Request and Response header.
  In this API if someone sets an HTTP header with a value we expect them to set a complete value in advance rather than appending it later to the header.
  Also using this API one can specify the header name in the list only once to avoid providing two entries of the same header name.

> - For headers like X-Forwarded-For it makes sense to add append action. So why does this enhancement not cover append ?
>   Answer: Currently, we are proceeding with Set and Delete only. The epic demands these two only. We have confirmed with the stakeholders as well.
>   Talking concretely about the X-Forwarded headers, we have a dedicated IngressController's and Route field to handle them which does have an `Append` action:

Reviewer 4: ```suggestion
- For headers such as `X-Forwarded-For`, it makes sense to add an "Append" action. So why does this enhancement not cover appending?
  Answer: Currently, we are proceeding with Set and Delete only. The RFE requires these two only. We have confirmed this with the stakeholders as well.
  For the `X-Forwarded-*` headers, we have a dedicated API field for IngressControllers and a Route annotation, which provide the equivalent of an `Append` action for these headers:
```

Author: done

```sh
$ oc explain ingresscontroller.spec.httpHeaders.forwardedHeaderPolicy
KIND:     IngressController
VERSION:  operator.openshift.io/v1

FIELD:    forwardedHeaderPolicy <string>

    DESCRIPTION:
    forwardedHeaderPolicy specifies when and how the IngressController sets the
    Forwarded, X-Forwarded-For, X-Forwarded-Host, X-Forwarded-Port,
    X-Forwarded-Proto, and X-Forwarded-Proto-Version HTTP headers. The value
    may be one of the following:
    * "Append", which specifies that the IngressController appends the headers,
    preserving existing headers.
    * "Replace", which specifies that the IngressController sets the headers,
    replacing any existing Forwarded or X-Forwarded-* headers.
    * "IfNone", which specifies that the IngressController sets the headers if
    they are not already set.
    * "Never", which specifies that the IngressController never sets the
    headers, preserving any existing headers. By default, the policy is
    "Append".
```

> - Would some headers be blocked for setting and deleting ?
>   Answer: Yes, this enhancement won't allow the user to set or delete HSTS headers. We already have a way to set this which is via route 
>   annotation. Providing one more way to add this can cause issues.

Reviewer 4: ```suggestion
- Are some headers blocked from being set or deleted?
  Answer: Yes, this enhancement won't allow the user to set or delete HSTS headers. We already have a way to set this via a Route 
  annotation. Providing another way to configure HSTS could cause issues.
```

- Many applications serve mixed content: static assets (images, js, css) and dynamic content (html, json). They usually want to cache the static assets on the client and do not care about session stickiness for those requests. Only the dynamic requests need to stick to the same server (session sticky).
Most applications will set something like this for static assets:
```sh
cache-control: max-age=7200, must-revalidate, public
```
But the OpenShift router (HAProxy) will add cache-control: private which results in both headers in the response:
```sh
cache-control: private
cache-control: max-age=7200, must-revalidate, public
```
This leads to issues with caching: CDNs like Cloudflare will not cache your files.

> Can we please make sure to only add the cache-control: private if there is no cache-control: public already present in the response ?
> Answer: A common technique is to put static assets on a different virtual host. In OpenShift terms, that means using a distinct route for static assets. This route can be annotated with haproxy.router.openshift.io/disable_cookies=true to prevent setting the cookie and thus prevent setting cache-control.

Reviewer 4: ```suggestion
Can we please make sure to only add the `cache-control: private` header if there is no `cache-control: public` header already present in the response?
Answer: A common technique is to put static assets on a different virtual host. In OpenShift terms, that means using a distinct route for static assets. This route can be annotated with `haproxy.router.openshift.io/disable_cookies=true` to prevent setting the cookie and thus prevent setting cache-control.
```

Author: done

- Would we be able to [Replace a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-a-header-by-using-a-regular-expression) or [Replace part of a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-part-of-a-header-by-using-a-regular-expression)?
  Answer: No, this is out of scope for this enhancement. Action `Set` will perform `set-header` in the haproxy.config for `http-response` and `http-request` which will either set the value if the value is not present or replace the existing value.

> - Can we be able to set dynamic header values ?

Reviewer 4: ```suggestion
- Should this enhancement allow setting dynamic header values?
```

Author: done

  Answer: Yes, you would be able to set a minimal set of dynamic header values. We had three options for this EP.
    1. Allow dynamic header values with all sample fetchers and converters.
    2. Allow a minimal set of dynamic values that we can confidently validate, focusing on the things that people have asked for on the RFE or [EP](https://github.com/openshift/enhancements/pull/1296#issuecomment-1396103783).
    3. Prohibit dynamic values entirely.
    However, as few customers asked for it, we went ahead with option number 2.

>     For setting dynamic header values for http response. Allowed sample fetchers are res.hdr, ssl_c_der. Allowed converters are lower, base64.
>     For setting dynamic header values for http request. Allowed sample fetchers are req.hdr, ssl_c_der. Allowed converters are lower, base64.

Reviewer 4: ```suggestion
    For setting dynamic header values for HTTP responses, allowed sample fetchers are `res.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
    For setting dynamic header values for HTTP requests, allowed sample fetchers are `req.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
```

Author: done

    Example: One might be able to set value of `x-forwarded-client-cert` to `%{+Q}[ssl_c_der,base64]`.

> - Certain headers like HSTS are not allowed to be set using this API as there is already a route annotation available for this

Reviewer 4: ```suggestion
- Certain headers like HSTS are not allowed to be set using this API as there is already a route annotation available for this.
```

Author: done

    problematically on another cluster. For these reasons, we decided not to implement the global-options-enable-hsts EP and instead wrote and implemented the
    global-admission-hsts EP, which defines a cluster-scoped API, using which the cluster admin can specify HSTS parameters that may or must be specified on routes
    for a given domain: https://github.com/openshift/api/blob/6ba31fa438f20d6c822ef47ad3d771309cb5216b/config/v1/types_ingress.go#L64-L87. Based on this experience,
    we wouldn't want to allow configuring HSTS on the IngressController using this new API.

### Test Plan

The controller that manages the IngressController Deployment and related
resources has unit test coverage; for this enhancement, the unit tests are
expanded to cover the additional functionality.

The operator has end-to-end tests; for this enhancement, a test is added that
(1) creates an IngressController that specifies certain response
headers should be set or deleted, (2) sends an HTTP request to one of the standard
routes (such as the console route), and (3) verifies that the expected headers

> are seen in the `curl -vk <route host>`.

Reviewer 4: This all looks reasonable.  You can also mention that integration tests in openshift/api, the unit tests in library-go, and the unit tests for the base64-encoding and quoting logic.

### Graduation Criteria

#### Dev Preview -> Tech Preview

#### Tech Preview -> GA

#### Removing a deprecated feature

### Upgrade / Downgrade Strategy

On upgrade, the default configuration sets or deletes no HTTP headers, as before the
introduction of the HTTP header Set/Delete API.

On downgrade, the HAProxy configuration template ignores unrecognized
environment variables set by ingress operator using the value set in the Ingress Controller CR and
the ServiceAlias config struct variable set by router using the value set in the routeSpec.

### Version Skew Strategy

### Operational Aspects of API Extensions

#### Failure Modes

#### Support Procedures

> 

Reviewer 4: These sections are still empty.  Hopefully we can avoid most of the possible failure modes through validation.  I suppose one failure mode would be exceeding `tune.bufsize`, or setting a header that confuses some server or some Web browser.  Support procedures could include checking access logs for client or server errors and checking for HAProxy config parse errors.

Author: somewhat github hid this comment not sure why. anyway will add these.

Author: tune.maxrewrite {{ env "ROUTER_MAX_REWRITE_SIZE" "8192" }}
  tune.bufsize {{ env "ROUTER_BUF_SIZE" "32768" }

If an HTTP request is larger than (tune.bufsize - tune.maxrewrite), haproxy will
return HTTP 400 (Bad Request) error. 
Similarly if an HTTP response is larger
than this size, haproxy will return HTTP 502 (Bad Gateway).

32768-8192 = 24576

But we restrict the header value to 4096 . Please correct me if I am wrong. 
RadDeg | x!
-- | --
Inv | sin | ln
π | cos | log
e | tan | √
Ans | EXP | xy

</div></div></div></div></div><div class="gjMOte" style="color: rgb(32, 33, 36); font-family: arial, sans-serif; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="KFFQ0c xKf9F" style="margin-top: 16px; align-items: center; display: flex; margin-bottom: 0px; width: 652px;"><br class="Apple-interchange-newline">

Reviewer 4: To tie up this loose end, the issue is that the user could set multiple headers, or set a header with a dynamic value that expanded to something very large, and thereby exceed the max rewrite buffer size.  Also, the cluster-admin could configure the max rewrite buffer to a smaller size than the default.  Because of dynamic values, the adjustable buffer size, and the multiple involved APIs (Route and IngressController), it is not practical for API validation to ensure that custom headers fit within the buffer.  This issue is now documented under "Failure Modes" in the EP, and the IngressController/Route API godoc for the `spec.httpHeaders.actions`/`value` fields mentions the issue as well.

   Example 3: As a user, I want to delete HTTP response headers such as `X-XSS-Protection` to prevent them from being sent to the client.
   Example 4: As a user I want to delete HTTP request headers such as `server` to prevent them from being sent to the application.

### Goals

1. Enable the cluster administrator to specify HTTP response headers in the IngressController CR spec that should be set or
   deleted at the global level when forwarding a response to all clients, i.e. for all routes.
2. Enable the cluster administrator to specify HTTP request headers in the IngressController CR spec that should be set or
   deleted at the global level when forwarding connections to all applications from clients.
3. Enable the route owner to specify HTTP response headers in the Route spec (when it is not a passthrough Route) for their application
   that should be set or deleted when forwarding a response to the client.
4. Enable the route owner to specify HTTP request headers in the Route spec (when it is not a passthrough Route) for their application
   that should be set or deleted when forwarding a connection to the application from a client.

> 5. Define an immutable/hard-coded deny list of headers which are not allowed to be set or deleted.

Reviewer 3: Is HSTS is the only entry in the deny list?  I feel like we would also need to include any of the headers that we specifically hard-code in the haproxy config template as I mentioned in:
https://github.com/openshift/router/pull/438/files#r1167340046
https://github.com/openshift/router/pull/438/files#r1167340360

But that is not the whole list.

    // httpHeaders defines policy for HTTP headers.
    //
    // If this field is empty, the default values are used.
    //
    // +optional
    HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`
}
```

```go
// IngressControllerHTTPHeaders specifies how the IngressController handles
// certain HTTP headers.
type IngressControllerHTTPHeaders struct {
    // actions specifies options for performing action on headers.
    // Note that this option only applies to cleartext HTTP connections
    // and to secure HTTP connections for which the ingress controller
    // terminates encryption (that is, edge-terminated or reencrypt
    // connections).  Headers cannot be set or deleted for TLS passthrough
    // connections.
    // Setting HSTS header is supported via another mechanism in Ingress.Spec.RequiredHSTSPolicies.
    // Setting httpCaptureHeaders for the headers set via this API will not work.
    // If spec.clientTLS, spec.httpHeaders.forwardedHeaderPolicy, spec.httpHeaders.uniqueId, spec.httpHeaders.headerNameCaseAdjustments is set
    // and if you set the same header via this API then that header will get replaced with the value specified via this API.

>     // If cache-control is set using this API then the existing value of cache-control will get replaced.

Reviewer 3: The list of headers that are not allowed to be edited needs to be clearly listed, or added to the deny list so that users will be informed of what they are not allowed to edit.

```go
type RouteSpec struct {
	...

    // httpHeaders defines policy for HTTP headers.
    //
    // If this field is empty, the default values are used.
    //
    // +optional
    HTTPHeaders RouteHTTPHeaders `json:"httpHeaders,omitempty" protobuf:"bytes,9,opt,name=httpHeaders"`
}	

```

```go
// RouteHTTPHeaders defines policy for HTTP headers.
type RouteHTTPHeaders struct {
    // actions specifies actions to take on headers.
    // Note that this option only applies to cleartext HTTP connections
    // and to secure HTTP connections for which the ingress controller
    // terminates encryption (that is, edge-terminated or reencrypt
    // connections).  Headers cannot be set or deleted for TLS passthrough
    // connections.
    // Setting HSTS header is supported via another mechanism in Ingress.Spec.RequiredHSTSPolicies.
    // Setting httpCaptureHeaders for the headers set via this API will not work.
    // If spec.clientTLS, spec.httpHeaders.forwardedHeaderPolicy, spec.httpHeaders.uniqueId, spec.httpHeaders.headerNameCaseAdjustments is set
    // and if you set the same header via this API then that header will get replaced with the value specified via this API.

>     // If cache-control is set using this API then the existing value of cache-control will get replaced.

Reviewer 3: Same comment regarding clarifying the list of denied headers.

	SetHTTPRequestHeaders              []SetHTTPHeader
	DeleteHTTPRequestHeaders           []DeleteHTTPHeader
}
```
```go
// SetHTTPHeader specifies an HTTP header that should be set.
type SetHTTPHeader struct {
	// Name specifies an HTTP header name.
	Name string

	// Value specifies the header value.
	Value string

	// Action specifies the action to be performed.
	Action string
}

type DeleteHTTPHeader struct {
	// Name specifies an HTTP header name.
	Name string
}
```

In this case the ordering of actions will be maintained as we will be having access to the RouteSpec inside router. So, we will run a loop
on either `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type` or `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type` which will have the desired
sequence.

C) library-go code will be having validations for checking Route API fields.
   We will be creating an immutable/hard-coded deny-list of HTTP response headers in [library-go](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141)

>    which won't be allowed to be set or deleted. For the initial enhancement, the HSTS `strict-transport-security` header name is included in this deny-list.

Reviewer 3: More details needed here about any other headers that should be in the deny-list, or else an explanation as to why it was determined that we 1) wouldn't notify users when a header wasn't allowed and 2) would allow users to override values we have explicitly hard-coded in the HAProxy template.

Author: > More details needed here about any other headers that should be in the deny-list, or else an explanation as to why it was determined that we 1) wouldn't notify users when a header wasn't allowed 

The API will throw an message if HSTS is added.

and 2) would allow users to override values we have explicitly hard-coded in the HAProxy template.

 The route owner does over-ride the headers set by admin however the scope for it will be at his application level only and it won't cause any issues cluster-wide as we don't set these in the front-end. However cluster-admin should carefully set the headers which get set in the front end sections for all routes. We can may be document a list of request headers not permitted to be set by  the admin.

Did you get a chance to look at this https://github.com/openshift/router/pull/438#discussion_r1170637633 ?

Author: as discussed the EP has been updated with the relevant documentation in the Non-goals section

   if it doesn't exist and remove before adding if it does.
6. Allowing sample fetchers other than `req.hdr`, `res.hdr`, and `ssl_c_der` or converters other than `lower` and `base64` in dynamic header values.
7. Providing a way to prohibit the Route owner from overriding the configuration for an IngressController if the Route and IngressController both set or delete a header with the same name.
8. The http request headers X-SSL, X-SSL-Client-Verify, X-SSL-Client-Serial, X-SSL-Client-Version, X-SSL-Client-SHA1, X-SSL-Client-DN, X-SSL-Client-CN, X-SSL-Issuer, X-SSL-Client-NotBefore, X-SSL-Client-NotAfter, X-SSL-Client-DER which are
   set by the haproxy template can be over-ridden or deleted as we don't blacklist them. This EP doesn't have the ability to disallow this. However, we don't recommend changing these values.
9. We explicitly delete Proxy header in the haproxy template. This header can be added using this API as it is not put in the blacklist. So this EP doesn't have the ability to disallow this. However, we don't recommend adding this header using this API.

> 10. The `HOST` header is made sure that it is in lowercase. Using this API one might be able change this to upper-case as this EP doesn't have the ability to disallow this. However, we don't recommend this.

Reviewer 3: Thanks for addressing this.  I see this as needing more details than just listing as non-goals.  I think you need to address why you chose **not** to add these headers to the same "deny list" that you added the HSTS header to.  Add these to the same section in which you discuss why you added the HSTS header.  

Also, the company prefers the phrases "deny list" and "approved list" rather than blacklist and whitelist because they have the same meaning to everyone, and a better technical connotation.

Author: done

   Only 64 items in request and response header list will be allowed.
   Minimum length of the header value would have to be greater than 0 and less than 4096.
   Sample fetchers for response dynamic header value allowed are "res.hdr" and "ssl_c_der".
   Sample fetchers for request dynamic header value allowed are "req.hdr" and "ssl_c_der".
   Converters allowed for request and response are "lower" and "base64".
   Actions like Set/Delete for headers can't be set for passthrough routes.

D) route/v1 of openshift/api will contain the validations using kubebuilder tags and CEL.
The validations will be same as mentioned in point C).

E) operator/v1/types_ingress.go of openshift/api will contain the validations using kubebuilder tags and CEL.
The validations will be same as mentioned in point C) except the one for passthrough routes.

F) To validate the header value it has to match a regex.

For the IngressController CR and Route CR (which is used for MicroShift), this validation is defined on the request and response action lists in openshift/api.

> For the Route API as handled by openshift-apiserver (which is used for OpenShift), this is defined in library-go.

Reviewer 3: @miheer @JoelSpeed @Miciah I still don't quite understand why the regexes need to be declared in two different places.  This seems like something that we would want to avoid in the design, and if we purposely chose to do this, a reason why it couldn't be done in a more design-first manner?

Author: @candita this epic can't do anything about this. We are going by the design which exists. The openshift-apiserver does not use kubebuilder tags for validations for route so we need library-go to get these validations done.

Author: @candita just started a thread here https://github.com/openshift/library-go/pull/1449#discussion_r1191208997

4. Enable the route owner to specify HTTP request headers in the Route spec (when it is not a passthrough Route) for their application
   that should be set or deleted when forwarding a connection to the application from a client.
5. Define a hard-coded deny-list of headers which are not allowed to be set or deleted (e.g. the `strict-transport-security` header;
   see [Risks and Mitigations](#risks-and-mitigations) for details).
6. Setting dynamic header values for HTTP responses. Allowed sample fetchers are `res.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
7. Setting dynamic header values for HTTP requests. Allowed sample fetchers are `req.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
8. Configuration for setting or deleting a header on a Route will override configuration for a header with the same name on the IngressController.

### Non-Goals

1. Validation that each specified value is valid for the respective header (e.g. require that either "DENY" or "SAMEORIGIN" is 
   specified for [the `x-frame-options` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options)).

> 2. Setting load balancing algorithm using header values. Please refer `Open Questions` for more details.

Author: done

> 2. Setting load balancing algorithm using header values. Please refer `Open Questions` for more details.

Reviewer 4: Now that you have sub-headings, you can link directly to the relevant question:
```markdown
2. Setting load balancing algorithm using header values. Please refer to [Open Questions](#will-this-enhancement-cover-setting-the-load-balancing-algorithm-based-on-headers) for more details.
```

3. Configuring buffer sizes for headers (e.g. customizing HAProxy's `tune.bufsize` and `tune.maxrewrite` settings).
4. Modifying headers for a passthrough route. 
5. Providing dedicated API knobs for [appending](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response%20add-header) or

>    [replacing](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response%20replace-header) headers..

Reviewer 4: You have an extra period.  
```suggestion
   [replacing](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response%20replace-header) headers.
```

Author: done

6. Allowing sample fetchers other than `req.hdr`, `res.hdr`, and `ssl_c_der` or converters other than `lower` and `base64` in dynamic header values.
   (Additional sample fetchers or converters may be allowed in the future, but not in the initial implementation of this EP.)

## Proposal

### Workflow Description

**Cluster admin** is a human user responsible for deploying a
cluster.

**Users** is a human user responsible for developing and

> deploying an application in a <cluster class=""></cluster>

Reviewer 4: ```suggestion
deploying an application in a cluster.
```

Author: not sure why goland added css

Author: done

    // and if you set the same header via this API then that header will get replaced with the value specified via this API.
    // If cache-control is set using this API then the existing value of cache-control will get replaced.
    // The header set using route will over-ride the header set in ingress controller if the headers defined both were same.
    // +optional
    Actions RouteHTTPHeaderActions `json:"actions,omitempty" protobuf:"bytes,1,opt,name=actions"`
}
```

```go
// RouteHTTPHeaderActions defines configuration for actions on HTTP request and response headers.
type RouteHTTPHeaderActions struct {
    // response is a list of HTTP response headers to set or delete.
    // The actions either Set or Delete will be performed on the headers in sequence as defined in the list of response headers.
    // A maximum of 64 response header actions may be configured.
    // You can use this field to specify HTTP response headers that should be set or deleted
    // when forwarding responses from your application to the client.
    // Note: This field cannot be used if your route uses TLS passthrough.
    // + ---

>     // + Note: Any change to regex mentioned below must be reflected in the crd validation of route in library-go(https://github.com/openshift/library-go/blob/master/pkg/route/validation/validation.go) or vice-versa.

Reviewer 4: ```suggestion
    // + Note: Any change to regex mentioned below must be reflected in the CRD validation of route in https://github.com/openshift/library-go/blob/master/pkg/route/validation/validation.go and vice-versa.
```

Author: done

```go
// RouteHTTPHeaderActionType defines actions that can be performed on HTTP headers.
type RouteHTTPHeaderActionType string

const (
    // Set specifies that an HTTP header should be set.
    Set RouteHTTPHeaderActionType = "Set"
    // Delete specifies that an HTTP header should be deleted.
    Delete RouteHTTPHeaderActionType = "Delete"
)
```

The `RouteSetHTTPHeader` type has a field for specifying a header value for the "Set" action.
```go
// RouteSetHTTPHeader specifies what value needs to be set on an HTTP header.
type RouteSetHTTPHeader struct {
    // value specifies a header value.
    // Dynamic values can be added. The value will be interpreted as an HAProxy format string as defined in
    // http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#8.2.6 and may use HAProxy's %[] syntax and
    // otherwise must be a valid HTTP header value as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2
    // Sample fetchers allowed are "req.hdr" and "ssl_c_der".
    // Converters allowed are "lower" and "base64".
    // Examples of header value - "%[req.hdr(host),lower]", "%[req.hdr(host),base64]", "{+Q}[ssl_c_der,base64]", "{+Q}[ssl_c_der,lower]"

>     // Note: Any change to regex mentioned below shall be reflected in the crd validation of route in library-go or vice-versa.

Author: Actually the regex is not here now.

but have actual use-cases (which aren't covered by the existing option). They will be allowed, with notes in the API and product documentation,
and possibly a warning on admission.
Eg: If someone creates a Ingress Controller with the following:
```sh
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
name: default
namespace: openshift-ingress-operator
spec:
httpHeaders:
forwardedHeaderPolicy: Replace
```
Then the `set-header` stanzas will be added in `haproxy.config` using the [haproxy-config.template](https://github.com/openshift/router/blob/0cd9a58dff77127562f7851ccfa5cda79b8cef48/images/router/haproxy/conf/haproxy-config.template#L565-L580)
However, if someone wants to set custom values for these then they need to use the API of this EP.
If both are set then custom values will take over.
Please note that the setting of forwardedHeaderPolicy happens at the plain backend section of haproxy.config i.e per route level and not at the front-end section
of the haproxy.config level as per above mentioned link.
Another use case is deleting the header which forwardedHeaderPolicy does not provide but this EP does.

#### X-SSL

> `x-ssl` and related headers similarly have an actual use-case, and allowing the cluster admin or route owner to modify these presents no security risk.

Reviewer 4: I think this is clearer:
```suggestion
Similarly to `x-forwarded-for`, there are legitimate use-cases for setting `x-ssl` and related headers, and allowing the cluster-admin or route owner to modify these headers presents no security risk.
```

Author: done

Eg: haproxy takes information from TLS handshake or tls session or the certificate that client provided. These headers are used for mtls and client certificates.
These headers tell you whether the client provided a certificate, was the certificate valid, what was the content or cipher, distinguished name or expiration date
on the certificate using converters with prefix like `ssl_` which can be used in the haproxy proxy config to take that information and put it http request headers which get sent

> to the backend application.

Reviewer 4: Please check the grammar and spelling again.  Change "a or b or c" to "a, b, or c".  Fix capitalization: "HAProxy", "TLS", "HTTP", "mTLS".  Actually, it would be better to spell out "mTLS" as "mutual TLS".   

On a related note, the IBM Style Guide says to use "for example" instead of "e.g.".

Author: done

Reviewer 4: Still not fixed: https://github.com/openshift/enhancements/blob/ae5a9e6428596afcad8ad1c83f801768480c308b/enhancements/ingress/set-delete-http-headers.md#L763

> You must already trust the cluster-admin or route owner who has full control over the proxy or application, so there is no additional risk in trusting
> an x-ssl header that is controlled by the cluster-admin or route owner which is the reason why we have not put this header in the deny list.

Reviewer 4: You usually use ``` `` ``` for header names, so please do so consistently.  I suggest using simpler, more explicit sentences for clarity.  
```suggestion
You must already trust the cluster-admin or route owner, who has full control over the proxy or application.  Thus there is no additional risk in trusting
an `x-ssl` header that is controlled by the cluster-admin or route owner.  For this reason, we do not restrict setting this header.
```

Author: done

#### Host

> Setting `Host` header via ingress controller CR can cause the map lookup to fail, and for that reason, we will prohibit setting the host request header on the IngressController.

Reviewer 4: ```suggestion
Setting the `host` header via the IngressController CR can cause HAProxy to fail to look up the correct route, and for that reason, we prohibit setting the `host` request header on the IngressController.
```

Author: done

    library-go validation on MicroShift that openshift-apiserver uses on OpenShift (see <https://github.com/openshift/kubernetes/pull/1416>).  As a result, the kubebuilder validation tags
    on the Route CRD are no longer required specifically for MicroShift; however, the kubebuilder validation tags are still required for client-side validation.

- Why don't you set/delete the headers on frontend public_ssl in the haproxy.config?
    Answer:
    The connection is still encrypted in public_ssl, so it is not possible to examine or modify the HTTP session there. The fe_sni and fe_no_sni frontends terminate TLS.

- Can one set load balancing algorithm using headers ?
    Answer:
    The HAproxy configuration provides a [load balancing algorithm](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-balance) which relies on the
    check against the incoming request's header values. Even though this enhancement adds the header manipulation capability it
    doesn't aim at providing (or enhancing) the load balancing API for the IngressController or Route.  Therefore, such a

>     load balancing algorithm is out of scope for this enhancement and should be handled with a separate RFE.

Reviewer 4: This looks like a duplicate of an earlier question: 

> Will this enhancement cover setting the load-balancing algorithm based on headers? Answer: No, this out of scope for this enhancement, so it will be considered as a non-goal. Please check point number 2 in the non-goals section for more details.

Please merge these two questions.

Author: done

- Unit tests to check the percent-encoded environment-variable values in the router deployment will be added in the ingress operator.
- Unit tests to check if values are passed correctly to the template and properly quoted in `haproxy.config` will be written in the router.
- Unit tests to test escaping will be added to library-go.

The operator has end-to-end tests; for this enhancement, a test is added that
(1) creates an IngressController that specifies certain response
headers should be set or deleted, (2) sends an HTTP request to one of the standard
routes (such as the console route), and (3) verifies that the expected headers
are seen in the `curl -vk <route host>`.

Integration tests will be added in openshift/api.

### Graduation Criteria

This feature will be GA from the start with no dev preview or tech preview.

#### Dev Preview -> Tech Preview

N/A

#### Tech Preview -> GA

N/A

#### Removing a deprecated feature

N/A

### Upgrade / Downgrade Strategy

On upgrade, the default configuration sets or deletes no HTTP headers, as before the
introduction of the HTTP header Set/Delete API.

> On downgrade, the httpHeaders API field won't be available for any HTTP requests or HTTP response to be set/deleted
> using the IngressController or Route Spec. So, if any existing values present in the IngressController or RouteSpec
> will be dropped by the api-server as the API fields won't be available in the lower version.

Reviewer 4: This is a little simpler:
```suggestion
On downgrade, the `httpHeaders` API field won't exist in the downgraded version, and so any header customizations that were specified in any IngressController or Route objects before the downgrade
will be dropped by the API server on downgrade and therefore will be  dropped by the router as well.
```

Author: done

2. If the ingress operator crashes after setting the httpHeaders in the IngressController Spec then check the ingress-operator logs using `oc logs <ingress-operator-pod-name> -n openshift-ingress-operator`.
3. If the router pods crash after setting the httpHeaders in the IngressController Spec then check the router pod logs using `oc logs <router pod name> -n openshift-ingress` and also check the ingress-operator pod logs .
4. If the router pods crash after setting the httpHeaders in the Route Spec then check the router pod logs using `oc logs <router pod name> -n openshift-ingress`.
5. In case of point 3 and 4 you may also run `oc debug deployment router-name -n openshift-ingress`, then check what `http-response/http-request set-header` or `http-response/http-request  del-header` stanzas were set in the haproxy.config
   If HTTP headers were set/delete via IngressController Spec then you need check the front end sections in the haproxy.config for these stanzas.
   If HTTP headers were set/delete via RouteSpec then you need to check the Plain backend section in the haproxy.config for these stanzas.

> 6. If you are not able to set the API fields via IngressController Spec or Route Spec then check the openshift-apiserver logs.

Author: @Miciah I have tried to address this in my PR. PTAL.

    Value string `json:"value"`
}
```
This alternative has the disadvantage that it relies more heavily on CEL and limits the extensibility of the API.  Unions are already pervasive in OpenShift APIs and facilitate
greater extensibility for adding more actions or action-specific options in the future.  For these reasons, we decided to use union types for this feature.

### Alternative quoting strategies for `haproxy.config`
We have multiple options for quoting header values in `haproxy.config`, as [described in the HAProxy documentation](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#2.2):
- Escape special characters (`\`, `'`, `"`, `#`, and `$`) using backslashes.
- Quote the value using double-quotes.
- Quote the value using single-quotes.
- For this EP we decided to quote using single-quotes, using the following steps:
  1. Replace every occurence of `'` with `'\''`.
  2. Wrap the entire string in single-quotes.
The HAProxy documentation refers to this as _strong quoting_.  Using strong quoting allows us to keep the quoting logic simple and `haproxy.config` more human-readable compared to the alternatives.

> ## Infrastructure Needed

Reviewer 4: ```suggestion
## Infrastructure Needed

This feature requires no special infrastructure.  
```

Author: done

---

> title: set-replace-append-delete-deny-headers

Reviewer 3: ```suggestion
title: set-delete-http-response-headers
```

authors:
  - "@miheer"
reviewers:
  - "@frobware"
  - "@gcs278"
  - "@Miciah"
approvers:
  - "@frobware"
  - "@gcs278"
  - "@Miciah"
api-approvers:
  - "@JoelSpeed"
  - "@deads2k"
  - "@frobware"
  - "@gcs278"
  - "@Miciah"
creation-date: 2022-12-02

> last-updated: 2023-03-26

Reviewer 3: Some date in July?

Author: done

tracking-link:
  - "https://issues.redhat.com/browse/NE-982"
  - "https://issues.redhat.com/browse/RFE-464"
status: TODO
see-also:
  - ""
replaces:
superseded-by:
---

# Set, Delete HTTP Response headers

## Summary

This enhancement extends the IngressController API and Route API to allow the user to set or delete
specific HTTP request and response headers (barring certain headers that it may be dangerous to allow customizing).

## Motivation

Certain HTTP request or response headers may need to be set or deleted to meet security compliance requirements, to prevent cross-site scripting attacks, to share
information about client connections or TLS sessions with backend servers, or to hide information about internal systems from clients.
In OpenShift 3, cluster administrators could use a custom HAProxy configuration template to set or delete headers, but using a custom
template is not permitted in OpenShift 4.

> Certain HTTP response headers like HSTS (Strict-Transport-Security) need to be denied as we already provide a way to set them
> via route annotation.

Reviewer 3: Please just delete this line.
```suggestion

```

Author: done

2. Setting load balancing algorithm using header values. Please refer [Open Questions](#open-questions) for more details.
3. Configuring buffer sizes for headers (for example: customizing HAProxy's `tune.bufsize` and `tune.maxrewrite` settings).
4. Modifying headers for a passthrough route. 
5. Providing dedicated API knobs for [appending](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response%20add-header) or
   [replacing](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response%20replace-header) headers.
6. Allowing sample fetchers other than `req.hdr`, `res.hdr`, and `ssl_c_der` or converters other than `lower` and `base64` in dynamic header values.
   (Additional sample fetchers or converters may be allowed in the future, but not in the initial implementation of this EP.)
7. Allowing conditions like `if` to set header values. For example: `x-frame-options: "%[res.hdr(server)] if { req.hdr(user-agent) -m sub evil}"` 
   Please refer [Open Questions](#open-questions) for more details.

## Proposal

### Workflow Description

**Cluster admin** is a human user responsible for deploying a
cluster.

> **Users** is a human user responsible for developing and

Reviewer 3: ```suggestion
**User** is a human user responsible for developing and
```

Author: done

                set:
                  value: /my-first-blog-post
            - name: X-SSL-Client-Cert
              action:
                type: Set
                set:
                  value: "%{+Q}[ssl_c_der,base64]"
            - name:   Content-Language
              action:
                type: Delete
            - name: X-Target
              action:
                type: Set
                set:
                  value: "%[req.hdr(host),lower]"
```

### API Extensions

A)  Setting or deleting HTTP request or response headers via `IngressController.spec`.
The existing `IngressControllerHTTPHeaders` Go type definition of the IngressController API is extended by adding a field for header manipulations.
```go
// IngressControllerSpec is the specification of the desired behavior of the
// IngressController.
type IngressControllerSpec struct {
    // httpHeaders defines policy for HTTP headers.
    //
    // If this field is empty, the default values are used.
    //
    // +optional
    HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`
}
```

```go
// IngressControllerHTTPHeaders specifies how the IngressController handles
// certain HTTP headers.

> type IngressControllerHTTPHeaders struct {

Reviewer 3: Can you update this from openshift/api?  it is called `IngressControllerHTTPHeaderActions` there.

> type IngressControllerHTTPHeaders struct {

Reviewer 3: Usually we put a ... to indicate something was omitted from the original. Without it, this looks like a new type here.

Author: done

>     // actions specifies options for performing action on headers.

Reviewer 3: Usually we put a `...` to indicate something was omitted from the original.  Without it, this looks like a new type here.

    // Note that this option only applies to cleartext HTTP connections
    // and to secure HTTP connections for which the ingress controller
    // terminates encryption (that is, edge-terminated or reencrypt
    // connections).  Headers cannot be set or deleted for TLS passthrough
    // connections.

>     // Setting HSTS header is supported via another mechanism in Ingress.Spec.RequiredHSTSPolicies.
>     // Setting httpCaptureHeaders for the headers set via this API will not work.
>     // If spec.clientTLS, spec.httpHeaders.forwardedHeaderPolicy, spec.httpHeaders.uniqueId, spec.httpHeaders.headerNameCaseAdjustments is set
>     // and if you set the same header via this API then that header will get replaced with the value specified via this API.
>     // If cache-control is set using this API then the existing value of cache-control will get replaced.
>     // The header set using route will over-ride the header set in ingress controller if the headers defined both were same.

Reviewer 3: ```suggestion
    // Setting HSTS header is supported via another mechanism in Ingress.Spec.RequiredHSTSPolicies.
    // The header settings made here can replace settings made in: cache-control, spec.clientTLS, 
    // spec.httpHeaders.forwardedHeaderPolicy, spec.httpHeaders.uniqueId, 
    // and spec.httpHeaders.headerNameCaseAdjustments.
    // The header settings made here can be replaced by Route annotations made by users.
```

Author: done

    // +optional
    Actions IngressControllerHTTPHeaderActions `json:"actions,omitempty"`
}
```

```go
// IngressControllerHTTPHeaderActions defines configuration for actions on HTTP request and response headers.
type IngressControllerHTTPHeaderActions struct {

>     // response is a list of HTTP response headers to set or delete.
>     // The actions either Set or Delete will be performed on the headers in sequence as defined in the list of response headers.
>     // A maximum of 64 response header actions may be configured.
>     // Enable the cluster administrator to specify HTTP response headers in the IngressController CR spec that should be set or
>     // deleted at the global level when forwarding responses from an application to the client, i.e. for all routes.

Reviewer 3: ```suggestion
    // response is a list of HTTP response headers to set or delete.
    // The cluster administrator uses these to specify HTTP response headers that should be set or
    // deleted at the global level when forwarding responses from an application to the client, i.e. for all routes.  
    // The actions either Set or Delete will be performed on the headers in sequence as defined in the list of response headers.
    // A maximum of 64 response header actions may be configured.
  
```

Author: done

    // + Note: Any change to regex mentioned below must be reflected in the CRD validation of route in https://github.com/openshift/library-go/blob/master/pkg/route/validation/validation.go and vice-versa.
    // +listType=map
    // +listMapKey=name
    // +optional
    // +kubebuilder:validation:MaxItems=64
    // +kubebuilder:validation:XValidation:rule=`self.all(key, key.action.type == "Delete" || (has(key.action.set) && key.action.set.value.matches('^(?:%(?:%|(?:\\{[-+]?[QXE](,[-+]?[QXE])*\\})?\\[(?:res\\.hdr\\([0-9A-Za-z-]+\\)|ssl_c_der)(?:,(?:lower|base64))*\\])|[^%[:cntrl:]]*)$')))`,
message="Either the header value provided is not in correct format or the sample fetcher/converter specified is not allowed. The dynamic header value will be interpreted as
an HAProxy format string as defined in http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#8.2.6 and may use HAProxy's %[] syntax and otherwise must be a valid HTTP
header value as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2 Sample fetchers allowed are res.hdr, ssl_c_der. Converters allowed are lower, base64."
    Response []IngressControllerHTTPHeader `json:"response,omitempty"`

>     // request is a list of HTTP request headers to set or delete.
>     // The actions either Set or Delete will be performed on the headers in sequence as defined in the list of request headers.
>     // A maximum of 64 request header actions may be configured.
>     // Enable the cluster administrator to specify HTTP request headers in the IngressController CR spec that should be set or
>     // deleted at the global level when forwarding responses to your application from the client, i.e. all routes.

Reviewer 3: ```suggestion
    // request is a list of HTTP request headers to set or delete.
     // The cluster administrator uses these to specify HTTP request headers that should be set or
    // deleted at the global level when forwarding responses to your application to the client, i.e. for all routes.  
    // The actions either Set or Delete will be performed on the headers in sequence as defined in the list of request headers.
    // A maximum of 64 request header actions may be configured.
    ```

Author: done

    // +listType=map
    // +listMapKey=name
    // +optional
    // +kubebuilder:validation:MaxItems=64
    // +kubebuilder:validation:XValidation:rule=`self.all(key, key.action.type == "Delete" || (has(key.action.set) && key.action.set.value.matches('^(?:%(?:%|(?:\\{[-+]?[QXE](,[-+]?[QXE])*\\})?\\[(?:req\\.hdr\\([0-9A-Za-z-]+\\)|ssl_c_der)(?:,(?:lower|base64))*\\])|[^%[:cntrl:]]*)$')))`,
message="Either the header value provided is not in correct format or the sample fetcher/converter specified is not allowed. The dynamic header value will be interpreted as an HAProxy
format string as defined in http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#8.2.6 and may use HAProxy's %[] syntax and otherwise must be a valid HTTP header value as
defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2 Sample fetchers allowed are req.hdr, ssl_c_der. Converters allowed are lower, base64."
    Request []IngressControllerHTTPHeader `json:"request,omitempty"`
}
```

```go
// IngressControllerHTTPHeader specifies configuration for setting or deleting an HTTP header.
type IngressControllerHTTPHeader struct {

>     // name specifies the name of a header to set or delete.  Its value must be a valid HTTP header.
>     // name as defined in RFC 2616 section 4.2.

Reviewer 3: ```suggestion
    // name specifies the name of a header to set or delete.  Its value must be a valid HTTP header
    // name as defined in RFC 2616 section 4.2.
```

Author: done

    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=256
    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
    // +kubebuilder:validation:XValidation:rule="self.lowerAscii() != 'strict-transport-security'",message="strict-transport-security may not be set via header actions"
    // +kubebuilder:validation:XValidation:rule="self.lowerAscii() != 'proxy'",message="proxy may not be set/delete via header actions"
    // +kubebuilder:validation:XValidation:rule="self.lowerAscii() != 'host'",message="host may not be set/delete via header actions"
    Name string `json:"name"`

    // action specifies actions to perform on headers, such as setting or deleting headers.
    // +kubebuilder:validation:Required
    Action IngressControllerHTTPHeaderActionUnion `json:"action"`
}
```

```go
// IngressControllerHTTPHeaderActionUnion specifies an action to take on an HTTP header.
// +union
type IngressControllerHTTPHeaderActionUnion struct {
    // type defines the type of the action to be applied on the header.

>     // Possible values are Set and Delete.
>     // Set allows you to set http request and response headers.
>     // Delete allows you to delete http request and response headers

Reviewer 3: ```suggestion
    // Possible values are Set or Delete.
    // Set allows you to set http request and response headers.
    // Delete allows you to delete http request and response headers.
```

Author: done

    // +unionDiscriminator
    // +kubebuilder:validation:Enum:=Set;Delete
    // +kubebuilder:validation:Required
    Type IngressControllerHTTPHeaderActionType `json:"type"`

    // set specifies how the HTTP header should be set.
    // If this field is empty, no action is taken.
    // +optional
    // +unionMember

>     Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`

Reviewer 3: This is clunky.  I don't know why IngressControllerSetHTTPHeader.value was not just put here and named Value.  If it's too late to change, then it's fine.

}
```

```go
// IngressControllerHTTPHeaderActionType defines actions that can be performed on HTTP headers.
type IngressControllerHTTPHeaderActionType string

const (
    // Set specifies that an HTTP header should be set.
    Set IngressControllerHTTPHeaderActionType = "Set"
    // Delete specifies that an HTTP header should be deleted.
    Delete IngressControllerHTTPHeaderActionType = "Delete"
)
```

The `IngressControllerSetHTTPHeader` type has a field for specifying a header value.
```go
// IngressControllerSetHTTPHeader defines the value which needs to be set on an HTTP header.
type IngressControllerSetHTTPHeader struct {
    // value specifies a header value.
    // Dynamic values can be added. The value will be interpreted as an HAProxy format string as defined in
    // http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#8.2.6 and may use HAProxy's %[] syntax and

>     // otherwise must be a valid HTTP header value as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2
>     // Sample fetchers allowed are "req.hdr" and "ssl_c_der".
>     // Converters allowed are "lower" and "base64".
>     // Examples of header value - "%[req.hdr(host),lower]", "%[req.hdr(host),base64]", "{+Q}[ssl_c_der,base64]", "{+Q}[ssl_c_der,lower]"

Reviewer 3: ```suggestion
    // otherwise must be a valid HTTP header value as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.
    // Sample fetchers allowed are "req.hdr" and "ssl_c_der".
    // Converters allowed are "lower" and "base64".
    // Examples of header value - "%[req.hdr(host),lower]", "%[req.hdr(host),base64]", "{+Q}[ssl_c_der,base64]", "{+Q}[ssl_c_der,lower]".
```

Author: ok done

    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=4096
    Value string `json:"value"`
}
```

B) Setting or deleting HTTP request or response headers via `Route.spec`.
```go
type Route struct {
...
    // spec is the desired state of the route

>     // +kubebuilder:validation:XValidation:rule="!has(self.tls) || self.tls.termination != 'passthrough' || !has(self.httpHeaders)",message="header actions are not permitted when tls termination is passthrough."

Reviewer 3: https://github.com/openshift/api/pull/1364/files#r1255016441

Author: your suggestion won't work as mentioned there

Author: That will fail.

```go
type RouteSpec struct {
	...

    // httpHeaders defines policy for HTTP headers.
    //
    // If this field is empty, the default values are used.
    //
    // +optional
    HTTPHeaders RouteHTTPHeaders `json:"httpHeaders,omitempty" protobuf:"bytes,9,opt,name=httpHeaders"`
}	

```

```go
// RouteHTTPHeaders defines policy for HTTP headers.
type RouteHTTPHeaders struct {
    // actions specifies actions to take on headers.
    // Note that this option only applies to cleartext HTTP connections
    // and to secure HTTP connections for which the ingress controller
    // terminates encryption (that is, edge-terminated or reencrypt
    // connections).  Headers cannot be set or deleted for TLS passthrough
    // connections.
    // Setting HSTS header is supported via another mechanism in Ingress.Spec.RequiredHSTSPolicies.
    // Setting httpCaptureHeaders for the headers set via this API will not work.
    // If spec.clientTLS, spec.httpHeaders.forwardedHeaderPolicy, spec.httpHeaders.uniqueId, spec.httpHeaders.headerNameCaseAdjustments is set
    // and if you set the same header via this API then that header will get replaced with the value specified via this API.

>     // If cache-control is set using this API then the existing value of cache-control will get replaced.

Reviewer 3: I think this doesn't apply for Route.

Author: yes. thanks!

HAProxy format string as defined in http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#8.2.6 and may use HAProxy's %[] syntax and otherwise must be a valid HTTP header value
as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2 Sample fetchers allowed are req.hdr, ssl_c_der. Converters allowed are lower, base64."
    Request []RouteHTTPHeader `json:"request,omitempty" protobuf:"bytes,2,opt,name=request"`
}
```

```go
// RouteHTTPHeader specifies configuration for setting or deleting an HTTP header.
type RouteHTTPHeader struct {
    // name specifies a header name to be set or deleted.  Its value must be a valid HTTP header.
    // name as defined in RFC 2616 section 4.2.
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=256
    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
    // +kubebuilder:validation:XValidation:rule="self.lowerAscii() != 'strict-transport-security'",message="strict-transport-security may not be set via header actions"

>     // +kubebuilder:validation:XValidation:rule="self.lowerAscii() != 'proxy'",message="proxy may not be set/delete via header actions"

Reviewer 3: Is the `host` header allowed for Routes?

Author: yes

    on the IngressController in the past; see the global-options-enable-hsts EP, which was never implemented. However, HSTS isn't neatly scoped: Route A's HSTS
    policy can affect Route B, even if the routes are in different namespaces or belong to different IngressControllers, if Route B's host name is a subdomain of
    Route A's. Additionally, specifying HSTS on an IngressController can cause a route manifest that behaves as expected on one cluster to behave surprisingly and
    problematically on another cluster. For these reasons, we decided not to implement the global-options-enable-hsts EP and instead wrote and implemented the
    global-admission-hsts EP, which defines a cluster-scoped API, using which the cluster admin can specify HSTS parameters that may or must be specified on routes
    for a given domain: https://github.com/openshift/api/blob/6ba31fa438f20d6c822ef47ad3d771309cb5216b/config/v1/types_ingress.go#L64-L87. Based on this experience,
    we wouldn't want to allow configuring HSTS on the IngressController using this new API.

> - Why do we need the kubebuilder tags and CEL on the route CR in openshift/api to perform validations on Microshift? Why do we need to add validations in library-go code
>   for route CR to perform validations on OCP?
>    Answer:
>     The requests coming into KAS go first through the kube-aggregator component, which determines whether the request should be handled locally or delegated to another apiserver.
>     On OCP, a route/v1 request will be delegated to OAS, where it will be decoded, defaulted, and then converted from the route/v1 version defined in o/api into the internal version
>     defined inside OAS. When it gets to OAS admission plugins, they're operating on the internal api [types](https://github.com/openshift/openshift-apiserver/blob/fc7db009c2c48eacba8bb1f21714[…]route/apiserver/admission/requiredrouteannotations/admission.go).
>     The original OAS behaviors are modified to operate on v1 instead of internal types and moved it into to library-go. None of the spec.host generation/authz, validation,
>     and field defaulting in OAS happen during admission, and that's still the case. They just have shims to convert internal->V1, call the shared library behavior, and
>     convert back V1->internal. In KAS, we don't have the same extension points since we're using the same CR handler that all CRs use. So in KAS, these same behaviors are performed by
>     admission plugins.
>     https://speakerdeck.com/sttts/kubernetes-api-codebase-tour?slide=19 does a good job illustrating why KAS route admission plugins are inert on OCP, but functional for MicroShift.
>     KAS admission plugins aren't invoked on OCP. On microshift, the kube-aggregator dispatches route requests to the in-process apiserver.
>     KAS admission route plugins are inert on OCP because routes never pass through the KAS admission chain there. They're proxied to OAS before that stage.
>     On MicroShift, that's not true and KAS admits+handles the route locally.
>     It was considered to be a good idea to take a similar course i.e. adding a new package under https://github.com/openshift/library-go/tree/master/pkg/route and
>     then adding a new admission plugin in openshift-apiserver for OCP and a new admission plugin for MicroShift, both using the library-go code with new api behaviors.
>     So, shared logic in library-go written against route/v1 types, KAS plugin converting unstructured->routev1, and OAS plugin converting internal->v1.
>     Even if something can be represented by a crd schema, sharing the implementation this way was going to remove the opportunity to introduce subtle differences between OAS-routes and CRD-routes.
>     So, in short it was earlier decided to use add validations(Eg: for an immutable list of prohibited values or combinations of values eg: HSTS) in library-go or admissions(Eg: for a dynamic list of prohibited values where
>     validation of one resource depends on data from another resource) using admission plugins in openshift/openshift-apiserver and openshift/kubernetes for OCP and Microshift respectively.
>     So this would avoid duplication for performing validations using library-go in 2 different places for the same route CR. However, there still will be duplication as we would need same consistent
>     admission plugins if needed to be written in openshift/kubernetes for Microshift and openshift/openshift-apiserver for OCP.
>     However later it was decided from a discussion with API team that route CR validations for OCP to be made in library-go code and for Microshift in openshift/api which should be consistent and well-formed.
>     The reason was just using CEL validations on CRD it is possible to express a wide range of previously impossible rules. The CRD schema provides help to client-side tools and interoperates with future schema checks for compatibility in progress.

Reviewer 4: @candita, @miheer, is this clearer?
```suggestion
- Why do we need the kubebuilder validation tags on the Route CRD in openshift/api in addition to validation in library-go?
   Answer:
    The library-go validation is used for server-side validation whereas the OpenAPI schema on the Route CRD may be used for client-side validation.  
    On OCP, kube-apiserver delegates the Route API to openshift-apiserver, which employs the library-go validation to validate Route objects (see <https://github.com/openshift/openshift-apiserver/blob/aac3dd5bf0547e928103a0f718ca104b1bb13930/pkg/route/apis/route/validation/validation.go>).  
    MicroShift does not use openshift-apiserver; instead, on MicroShift, kube-apiserver handles the Route API using the Route CRD.  However, kube-apiserver has an admission plugin for the Route CRD that employs the same library-go validation that openshift-apiserver uses (see <https://github.com/openshift/kubernetes/blob/15f19ea2dd700767e5337502aec753d2a6e26905/openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go>).
    Defining validation for the same API in both library-go and the OpenAPI schema introduces the risk that someone could inadvertently introduce discrepancies between the library-go validation and the OpenAPI schema, leading to buggy and potentially confusing behavior.  Care must be taken to avoid introducing such discrepancies.  
    Historical note: At one time, MicroShift relied on the OpenAPI schema for server-side validation of the Route CRD, and for this reason, kubebuilder validation tags were added to the Route CRD (see <https://github.com/openshift/api/pull/1228>).  Since then, an admission plugin has been added to kube-apiserver that validates the Route CRD using the same library-go validation on MicroShift that openshift-apiserver uses on OpenShift (see <https://github.com/openshift/kubernetes/pull/1416>).  As a result, the kubebuilder validation tags on the Route CRD are no longer required specifically for MicroShift; however, the kubebuilder validation tags are still required for client-side validation.  
```

Author: looks good. I made the changes

```

Allowed dynamic values for request would be -
```conf
%[req.hdr(host),lower]
%[req.hdr(host),base64]
%[req.hdr(Host)] if foo
```
Allowed dynamic values for response would be -
```conf
%[res.hdr(X-Acl-Key),lower]
%[res.hdr(X-Acl-Key),base64]
%[req.hdr(Host)] if foo
```
Dynamic values allowed for both request and response would be -
```conf
%{+Q}[ssl_c_der,base64]
```

For regular header which are not dynamic one can pass values as mentioned in unit test `TestHeaderValues` with `validInput:true`
present in `pkg/route/validation/validation_test.go`

Please note that all values will be quoted before interpolating them into `haproxy.conf`.  Specifically, any single-quote present in the value is escaped, and then the whole value
is surrounded in single-quotes.  This prevents HAProxy config parsing errors in the event of invalid input. This quoting logic will be present in the router.
To quote using single-quotes, we will do the following:
Replace ' with '\''.
Wrap the entire string in single quotes.
eg: "abc"\\ 'def'  shall it be converted to '"abc"\\ '\''def'\'''

### Risks and Mitigations

> We identified the following headers where some of them either can be allowed to set/delete whereas some headers need not to be allowed to set/delete
> due to the reasons mentioned in the following:

Reviewer 3: ```suggestion
We identified the following headers that require special explanations:
```

Author: ok i just wrote what we discussed on the call.

Author: done.

#### Proxy header
  As per haproxy security vulnerability, if the client sets `proxy` header, then the cgi app takes the value and injects it in the env var `HTTP_PROXY`,
  so basically client is setting the proxy which is a security vulnerability so that is why we deny setting/deleting proxy header with this EP.
  Please note that proxy headers are deleted by haproxy template by default.
  `proxy` isn't necessarily a security risk (route owners could inject configuration into their own applications) but is non-standard and potentially

>   confusing, so it should be blocked.

Author: It is not a CGI/PHP application vulnerability. I will try to reword it.

>   confusing, so it should be blocked.

Author: reworded it.

#### Strict-Transport-Security

>   strict-transport-security is a security risk (if we allowed it, then route owners could impact each other's routes), so it must be blocked so we have put this header in the deny list.

Reviewer 3: ```suggestion
  Strict-transport-security is already handled separately and we don't want to allow overrides of that configuration. If we allowed it, then route owners could impact each other's routes. It must be blocked so we have put this header in the deny list.
```

Author: done

  Please refer the question `Certain headers like HSTS are not allowed to be set using this API as there is already a route annotation available for this.
  Would it not be better to provide a concrete API and deprecate the annotation API?` in the `Open Questions` section.

#### X-Forwarded-For

>   `x-forwarded-for` and related headers are potentially confusing (as modifying these headers using this new API could affect configuration specified using
>    another [option](https://docs.openshift.com/container-platform/4.12/networking/ingress-operator.html#nw-using-ingress-forwarded_configuring-ingress))
>   but have actual use-cases (which aren't covered by the existing option), so they should be allowed, with notes in the API and product documentation,
>   and possibly a warning on admission.

Reviewer 3: ```suggestion
  `x-forwarded-for` and related headers were considered for the deny-list (as modifying these headers using this new API could affect configuration specified using
   another [option](https://docs.openshift.com/container-platform/4.12/networking/ingress-operator.html#nw-using-ingress-forwarded_configuring-ingress))
  but have actual use-cases (which aren't covered by the existing option).   They will be allowed, with notes in the API and product documentation,
  and possibly a warning on admission.
```

Reviewer 3: It will be helpful to list an example of such a use case.

#### X-SSL
 `x-ssl` and related headers similarly have an actual use-case, and allowing the cluster admin or route owner to modify these presents no security risk.
  Eg: haproxy takes information from tls handshake or tls session or the certificate that client provided. These headers are used for mtls and client certificates.
  These headers tell you did the client provide a certificate or was the certificate valid or what was the content or cipher or distinguished name or expiration date
  on the certificate using converters with prefix like `ssl_` which can be used in the haproxy proxy config to take that information and put it http request headers which get sent

>   to the backend application.

Reviewer 3: This is not clear, needs some re-wording and a grammar check.  If I could figure it out, I'd make a suggestion, but I can't.

Author: I am not sure how to reword it more. If you can explain what is needed then that will be great.

Author: just made some grammar changes

- Should this enhancement allow setting dynamic header values?
  Answer: Yes, you would be able to set a minimal set of dynamic header values. We had three options for this EP.
    1. Allow dynamic header values with all sample fetchers and converters.
    2. Allow a minimal set of dynamic values that we can confidently validate, focusing on the things that people have asked for on the RFE or [EP](https://github.com/openshift/enhancements/pull/1296#issuecomment-1396103783).
    3. Prohibit dynamic values entirely.
    However, as few customers asked for it, we went ahead with option number 2.
    For setting dynamic header values for HTTP responses, allowed sample fetchers are `res.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
    For setting dynamic header values for HTTP requests, allowed sample fetchers are `req.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
    Example: One might be able to set value of `x-forwarded-client-cert` to `%{+Q}[ssl_c_der,base64]`.

> - Certain headers like HSTS are not allowed to be set using this API as there is already a route annotation available for this.

Reviewer 3: Again, HSTS is not just a route annotation.  It is a setting that an admin can add to a domain, via `spec.requiredHSTSPolicies`, as explained in https://docs.openshift.com/container-platform/4.13/networking/routes/route-configuration.html#nw-enforcing-hsts-per-domain_route-configuration.

  Would it not be better to provide a concrete API and deprecate the annotation API?
  Answer: The strict-transport-security header (AKA HSTS) has wider ranging implications than the typical HTTP header. We considered allowing HSTS to be configured

>     on the IngressController in the past; see the global-options-enable-hsts EP, which was never implemented. However, HSTS isn't neatly scoped: Route A's HSTS

Author: I have tried to modify this. PTAL.

---
title: set-replace-append-delete-deny-headers
authors:
  - "@miheer"
reviewers:
  - "@frobware"
  - "@gcs278"
  - "@Miciah"
approvers:
  - "@frobware"
  - "@gcs278"
  - "@Miciah"
api-approvers:
  - "@JoelSpeed"
  - "@deads2k"
  - "@frobware"
  - "@gcs278"
  - "@Miciah"
creation-date: 2022-12-02
last-updated: 2023-03-26
tracking-link:
  - "https://issues.redhat.com/browse/NE-982"
  - "https://issues.redhat.com/browse/RFE-464"
status: TODO
see-also:
  - ""
replaces:
superseded-by:
---

# Set, Delete HTTP Response headers

## Summary

This enhancement extends the IngressController API and Route API to allow the user to set or delete

> specific HTTP request and response headers(barring certain headers that it may be dangerous to allow customizing).

Author: done

> specific HTTP request and response headers(barring certain headers that it may be dangerous to allow customizing).

Author: done

   Example 3: As a cluster administrator I want to delete HTTP response headers such as `server` to prevent them from being sent to the client.
   Example 4: As a cluster administrator I want to delete http request headers like `Content-Type` to be not sent to the application.
2. As a route owner, I need to configure my Route to set or delete HTTP request or response headers to send or not send them to the client which has connected to the Route.
   Example 1: As a user, I want to set HTTP response headers such as `X-XSS-Protection: 1; mode=block`, `X-Content-Type-Options: nosniff`, and `X-Frame-Options: SAMEORIGIN` to be sent to the client.
   Example 2: As a user, I want to set HTTP request headers such as `X-SSL-Client-Cert: %{+Q}[ssl_c_der,base64]` and `X-Target: %[req.hdr(host),lower]` (using HAProxy's  dynamic value syntax) to be sent to the application.
   Example 3: As a user, I want to delete HTTP response headers such as `X-XSS-Protection` to prevent them from being sent to the client.

>    Example 4: As a user I want to delete HTTP request headers such as `server` to prevent them from being sent to the application.

Author: ok done

>    Example 4: As a user I want to delete HTTP request headers such as `server` to prevent them from being sent to the application.

Reviewer 4: Looks like you used a nested list, which isn't what I suggested, but it is an improvement.

   deleted at the global level when forwarding a response to all clients, i.e. for all routes.
2. Enable the cluster administrator to specify HTTP request headers in the IngressController CR spec that should be set or
   deleted at the global level when forwarding connections to all applications from clients.
3. Enable the route owner to specify HTTP response headers in the Route spec (when it is not a passthrough Route) for their application
   that should be set or deleted when forwarding a response to the client.
4. Enable the route owner to specify HTTP request headers in the Route spec (when it is not a passthrough Route) for their application
   that should be set or deleted when forwarding a connection to the application from a client.
5. Define an immutable/hard-coded deny list of headers which are not allowed to be set or deleted.
   Eg- we should prohibit specifying strict-transport-security using the new API, for the following reasons:
   HSTS is already configurable through a route annotation.
   We already have an API for restricting HSTS in the cluster config.

>    Adding a second way to configure HSTS would significantly complicate this enhancement.

Author: done

6. Setting dynamic header values for HTTP responses. Allowed sample fetchers are `res.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
7. Setting dynamic header values for HTTP requests. Allowed sample fetchers are `req.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
8. Configuration for setting or deleting a header on a Route will override configuration for a header with the same name on the IngressController.

### Non-Goals

1. Validation that each specified value is valid for the respective header (e.g. require that either "DENY" or "SAMEORIGIN" is 
   specified for [the `x-frame-options` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options)).
2. Setting load balancing algorithm using header values. The HAproxy configuration provides a [load balancing algorithm](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-balance) which relies on the
   check against the incoming request's header values. Even though this enhancement adds the header manipulation capability it
   doesn't aim at providing (or enhancing) the load balancing API for the IngressController or Route.  Therefore such a 

>    load balancing algorithm is out of scope for this enhancement and should be handled with a separate RFE.

Author: done

    // and if you set the same header via this API then that header will get replaced with the value specified via this API.
    // If cache-control is set using this API then the existing value of cache-control will get replaced.
    // The header set using route will over-ride the header set in ingress controller if the headers defined both were same.
    // +optional
    Actions IngressControllerHTTPHeaderActions `json:"actions,omitempty"`
}
```

```go
// IngressControllerHTTPHeaderActions defines configuration for actions on HTTP request and response headers.
type IngressControllerHTTPHeaderActions struct {
    // response is a list of HTTP response headers to set or delete.
    // The actions either Set or Delete will be performed on the headers in sequence as defined in the list of response headers.
    // A maximum of 64 response header actions may be configured.
    // Enable the cluster administrator to specify HTTP response headers in the IngressController CR spec that should be set or
    // deleted at the global level when forwarding responses from an application to the client, i.e. for all routes.
    // + ---

>     // + Note: Any change to regex mentioned below shall be reflected in the crd validation of route in library-go(https://github.com/openshift/library-go/blob/master/pkg/route/validation/validation.go) or vice-versa.

Author: done

this gets decoded into the two triples ("Accept", "text/plain, text/html", "Set") and ("Accept-Encoding", "Delete"); then the template adds the following stanzas to the `public`, `fe_no_sni`, and `fe_sni` frontends:
```conf
    http-response set-header X-Frame-Options 'DENY'
    http-response set-header X-XSS-Protection '1;mode=block'
    http-response set-header x-forwarded-client-cert '%{+Q}[ssl_c_der,base64]'
    http-response del-header X-Frame-Options
    http-response del-header X-XSS-Protection
    http-request set-header Accept 'text/plain, text/html'
    http-request del-header Accept-Encoding
```
Header values are quoted so that HAProxy does not throw an error while parsing the haproxy template for `http-response` and `http-request` stanzas
that include spaces or other special characters.
Any single quotes provided within the header value will be escaped and replaced with '\'' and then the value is encapsulated with single quotes ''.
Eg: Header value `"abc"\ 'def'` is quoted as `'"abc"\ '\''def'\'''`

> Header value `%[req.hdr(Host)] if foo` is quoted as '%[req.hdr(Host)] if foo'

Author: done

The following shows how it will get set in haproxy.config
```conf

> http-response set-header X-Frame-Options '\"abc\"\ "def"'

Author: done

http-response set-header X-Server '%[req.hdr(Host)] if foo'
```

> Use of a single environment variable `ROUTER_HTTP_RESPONSE_HEADERS` for HTTP response headers  and `ROUTER_HTTP_REQUEST_HEADERS`for HTTP request headers, respectively, for both the `Set` and `Delete` actions enables the template to preserve ordering.

Author: done

This feature's implementation will be similar to an existing feature called logging headers which was implemented in
[router](https://github.com/openshift/router/commit/40ee0ba472a4ac503f1e6c97d4464f559101c97c) and 
[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/commit/c5139cb07c3a99a8a3b18a6c9321310add51b075).

B) Setting or Deleting HTTP headers via Route's Spec by Users.

Implementing this enhancement requires changes in the following repositories:

* openshift/api
* openshift/router
* openshift/openshift-apiserver

The router configures HAProxy using a configuration template. The template is populated with data from `ServiceAliasConfig` values, where each `ServiceAliasConfig` represents a Route.

> The enhancement adds four fields in `ServiceAliasConfig`: `SetHTTPResponseHeaders`, `DeleteHTTPResponseHeaders`, `SetHTTPRequestHeaders`, and `DeleteHTTPRequestHeaders` for all combinations of the "Set" and "Delete" actions and request or response headers to be set or deleted in the

Author: done

sequence. If request and response headers are provided in the API then first we run loop through request headers and then response headers.

C) library-go code will be having validations for checking Route API fields for OCP.
   We will be creating an immutable/hard-coded deny-list of HTTP response headers in [library-go](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141)
   which won't be allowed to be set or deleted. For the initial enhancement, the HSTS `strict-transport-security` header name is included in this deny-list.
   Only 64 items in request and response header list will be allowed.
   Minimum length of the header value would have to be greater than 0 and less than 4096.
   Sample fetchers for response dynamic header value allowed are "res.hdr" and "ssl_c_der".
   Sample fetchers for request dynamic header value allowed are "req.hdr" and "ssl_c_der".
   Converters allowed for request and response are "lower" and "base64".
   Actions like Set/Delete for headers can't be set for passthrough routes.

> D) route/v1 of openshift/api will contain the validations using kubebuilder tags and CEL for checking Route API fields for Microshift.

Author: done

The validations will be same as mentioned in point C).

E) operator/v1/types_ingress.go of openshift/api will contain the validations using kubebuilder tags and CEL.
The validations will be same as mentioned in point C) except the one for passthrough routes.

F) To validate the header value it has to match a regex.

> For the IngressController CR and Route CR (which is used for MicroShift), this validation is defined on the request and response action lists in openshift/api.
> For the Route API as handled by openshift-apiserver (which is used for OpenShift), this is defined in library-go.

Reviewer 4: MicroShift no longer uses the Route CRD.
```suggestion
For the IngressController CRD and Route CRD, this validation is defined on the request and response action lists in openshift/api.
For the Route API as handled by openshift-apiserver, this is defined in library-go.
```

Author: done

```
```go
response header:
`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```

Allowed dynamic values for request would be -
```conf
%[req.hdr(host),lower]
%[req.hdr(host),base64]
%[req.hdr(Host)] if foo
```
Allowed dynamic values for response would be -
```conf
%[res.hdr(X-Acl-Key),lower]
%[res.hdr(X-Acl-Key),base64]
%[req.hdr(Host)] if foo
```
Dynamic values allowed for both request and response would be -
```conf
%{+Q}[ssl_c_der,base64]
```

For regular header which are not dynamic one can pass values as mentioned in unit test `TestHeaderValues` with `validInput:true`
present in `pkg/route/validation/validation_test.go`

Please note that all values will be quoted before interpolating them into `haproxy.conf`.  Specifically, any single-quote present in the value is escaped, and then the whole value
is surrounded in single-quotes.  This prevents HAProxy config parsing errors in the event of invalid input. This quoting logic will be present in the router.
To quote using single-quotes, we will do the following:

> Replace ' with '\''.
> Wrap the entire string in single quotes.
> eg: "abc"\\ 'def'  shall it be converted to '"abc"\\ '\''def'\'''

Reviewer 4: Use ``` `` ``` to quote syntax.  Use a numbered list for steps where order matters.
```suggestion
1. Replace `'` with `'\''`.
2. Wrap the entire string in single quotes.

For example, `"abc"\\ 'def'`  will be converted to `'"abc"\\ '\''def'\'''`.
```

Author: done

### Risks and Mitigations

We identified the following headers that require special explanations:

#### Proxy header

>   As per haproxy security vulnerability, if the client sets `proxy` header, then the cgi app takes the value and injects it in the env var `HTTP_PROXY`,
>   so basically client is setting the proxy which is a security vulnerability as it is modifies the proxy value of the
>   environment variable which may be used by other applications so that is why we deny setting/deleting proxy header with this EP.
>   Please note that proxy headers are deleted by haproxy template by default.
>   `proxy` isn't necessarily a security risk (route owners could inject configuration into their own applications) but is non-standard and potentially
>   confusing, so it should be blocked.

Reviewer 4: * I asked you to link to https://httpoxy.org/.  Is there some reason not to?  
* This is not an HAProxy security vulnerability; it's a vulnerability in CGI.
* Why do you indent the whole subsection?
* Avoid colloquialisms such as "so basically".  
* Be explicit here about what the EP does.  Does the EP block routes from setting the `proxy` header?
```suggestion
The `proxy` HTTP request header can be used to exploit vulnerable CGI applications using the [_httpoxy_](https://httpoxy.org/) vulnerability.
In this vulnerability, if the client sets the HTTP `proxy` request header and the proxy passes it through to the application, then the CGI application takes the header value and injects it in the `HTTP_PROXY` environment variable.
In this manner, an attacker could direct a vulnerable application to use a proxy of the attacker's choice.

Note that OpenShift router already mitigates this vulnerability by deleting any `proxy` header that the client specifies.

Allowing the cluster-admin or route owner to override the `proxy` HTTP header isn't necessarily a security risk; it would only allow admins or route owners to inject configuration into their own applications, which they already control.  However, the `proxy` header is non-standard, and allowing it to be configured is potentially
confusing.  For this reason, we do not allow setting the `proxy` header with this EP.
```

Author: ok

Author: I asked you to link to https://httpoxy.org/. Is there some reason not to? I forgot to do so. sorry.

Why do you indent the whole subsection? --- i thought it will look good.

Be explicit here about what the EP does. Does the EP block routes from setting the proxy header?
Yes.

Reviewer 4: Thanks!  My suggestion had some paragraph breaks, which you removed, and the resulting paragraph is rather long.  Could you add back some paragraph breaks?

#### Strict-Transport-Security
Strict-transport-security is already handled separately and we don't want to
allow overrides of that configuration. If we allowed it, then route owners
could impact each other's routes. It must be blocked so we have put this
header in the deny list.

#### X-Forwarded-For
  `x-forwarded-for` and related headers were considered for the deny-list (as modifying these headers using this new API could affect configuration specified using
   another [option](https://docs.openshift.com/container-platform/4.12/networking/ingress-operator.html#nw-using-ingress-forwarded_configuring-ingress))
  but have actual use-cases (which aren't covered by the existing option). They will be allowed, with notes in the API and product documentation,
  and possibly a warning on admission.
  Eg: If someone creates a Ingress Controller with the following:
```sh
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
name: default
namespace: openshift-ingress-operator
spec:
httpHeaders:
forwardedHeaderPolicy: Replace
```

>   Then the set-header stanzas will be added in the haproxy.config using the [haproxy-config.template](https://github.com/openshift/router/edit/master/images/router/haproxy/conf/haproxy-config.template#L565-L580)

Reviewer 4: Use permalinks when possible.
```suggestion
  Then the `set-header` stanzas will be added in `haproxy.config` using the [haproxy-config.template](https://github.com/openshift/router/blob/0cd9a58dff77127562f7851ccfa5cda79b8cef48/images/router/haproxy/conf/haproxy-config.template#L565-L580)
```

  in the plain backend section of the haproxy.config.
  However, if someone wants to set custom values for these then they need to use the API of this EP.

>   If both are set then custom values will takeover.

Reviewer 4: The word "takeover" is a noun; use "take over" for the verb.  
```suggestion
  If both are set then custom values will take over.
```

Author: done

  Please note that the setting of forwardedHeaderPolicy happens at the plain backend section of haproxy.config i.e per route level and not at the front-end section
  of the haproxy.config level as per above mentioned link.
  Another use case is deleting the header which forwardedHeaderPolicy does not provide but this EP does.

#### X-SSL
 `x-ssl` and related headers similarly have an actual use-case, and allowing the cluster admin or route owner to modify these presents no security risk.
  Eg: haproxy takes information from tls handshake or tls session or the certificate that client provided. These headers are used for mtls and client certificates.
  These headers tell you whether the client provided a certificate or was the certificate valid or what was the content or cipher or distinguished name or expiration date
  on the certificate using converters with prefix like `ssl_` which can be used in the haproxy proxy config to take that information and put it http request headers which get sent

>   to the backend application.

Reviewer 4: What you wrote doesn't seem to explain why there is no security risk.  I think the reason boils down to the fact that you must already trust the cluster-admin or route owner who has full control over the proxy or application, so there is no additional risk in trusting an `x-ssl` header that is controlled by the cluster-admin or route owner.  Is that the gist of it?  

Also, please check the grammar and spelling (change "tls" to "TLS", and change "a or b or c" to "a, b, or c").

Author: yes

> #### HOST

Reviewer 4: Why do you capitalize "HOST" instead of just writing "Host"?

Author: OK made the change.

  Setting `Host` header via ingress controller CR can cause the map lookup to fail, and for that reason, we will prohibit setting the host request header on the IngressController.
  The haproxy.config uses map stanzas containing base(host+path) in the frontend sections which uses host from request header passed by client. Manually changing this from ingress
  controller causes routes to fail with 503 error.
  So, `Host` header is in the deny list of headers to be set/deleted by ingress controller.

### Drawbacks

## Design Details

> ### Open Questions [optional]

Reviewer 4: Please use `####` subsection headers.

Reviewer 4: `### Open Questions` is correct.  I'm saying that under the "Open Questions" heading, you can use `####` headings for individual questions.

Author: done

Reviewer 4: This still isn't done.

Author: @Miciah questions are now headers.

    library-go validation on MicroShift that openshift-apiserver uses on OpenShift (see <https://github.com/openshift/kubernetes/pull/1416>).  As a result, the kubebuilder validation tags
    on the Route CRD are no longer required specifically for MicroShift; however, the kubebuilder validation tags are still required for client-side validation.

- Why don't you set/delete the headers on frontend public_ssl in the haproxy.config?
    Answer:
    The connection is still encrypted in public_ssl, so it is not possible to examine or modify the HTTP session there. The fe_sni and fe_no_sni frontends terminate TLS.

### Test Plan

The controller that manages the IngressController Deployment and related
resources has unit test coverage; for this enhancement, the unit tests are
expanded to cover the additional functionality.
Unit tests for testing regular expression to test invalid values will be written in the library-go.
Unit test to check the deployment env var having base64 encoded value of the router deployment will be written in the ingress operator.
Unit tests to check if the values are passed correctly to the template will be written in the router.

> Unit tests to test escaping and quoting will be written in the library-go.

Author: done

> Unit tests to test escaping and quoting will be written in the library-go.

Reviewer 4: (There's still no mention of base64 encoding in the alternatives section.)

The operator has end-to-end tests; for this enhancement, a test is added that
(1) creates an IngressController that specifies certain response
headers should be set or deleted, (2) sends an HTTP request to one of the standard
routes (such as the console route), and (3) verifies that the expected headers
are seen in the `curl -vk <route host>`.

Integration tests will be added in openshift/api.

### Graduation Criteria

#### Dev Preview -> Tech Preview

#### Tech Preview -> GA

#### Removing a deprecated feature

> 

Author: done

### Upgrade / Downgrade Strategy

On upgrade, the default configuration sets or deletes no HTTP headers, as before the
introduction of the HTTP header Set/Delete API.

On downgrade, the HAProxy configuration template ignores unrecognized
environment variables set by ingress operator using the value set in the Ingress Controller CR and

> the ServiceAlias config struct variable set by router using the value set in the routeSpec.

Reviewer 4: How can the `ServiceAliasConfig` struct provide unrecognized values to the template?  The `ServiceAliasConfig` struct and `haproxy-config.template` are defined in the same container, so the versions should always match.

Author: i added a different text.

	Name string `json:"name"`

	// value specifies fields to mention the header details and the action you want to perform on them.
	// +optional
	// +kubebuilder:validation:XValidation:rule="has(self.value) && self.action == 'Delete'", message="Value is not required while performing delete action."
	// +kubebuilder:validation:XValidation:rule="!has(self.value) && self.action == 'Set'", message="Value is required while performing set action."
	// +kubebuilder:validation:MinLength=1
	// +kubebuilder:validation:MaxLength=4096
	Value string `json:"value"`

	// action defines the type of the action to be applied on the header.
	// +unionDiscriminator
	// +kubebuilder:validation:Enum:=Set;Delete
	// +kubebuilder:validation:Required
	Action IngressControllerHTTPHeaderAction `json:"action,omitempty"`
}

// IngressControllerHTTPHeaderAction defines operation you can perform on headers.
type IngressControllerHTTPHeaderAction string

const (
	// Set is used specify set action per header
	Set IngressControllerHTTPHeaderAction = "Set"
	// Delete is used specify delete action per header
	Delete IngressControllerHTTPHeaderAction = "Delete"
)

> ```

Author: ok thanks! done

2. To sanitize the header value we have two options:
    - Escape using backslashes
       This is an alternate solution if ones wants to quotes using backslahes.
        - Replace \ with \\.
        - Replace ' with \'.
        - Replace " with \".
        - Replace # with \#.
        - Replace $ with \$.
    - Escape using single-quotes.
       For this EP we decided to quote using single quotes, as per the following:
        - Replace ' with '\''.
        - Wrap the entire string in single quotes.
        As per http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#2.2 we use strong quoting.
        Strong quoting is achieved by surrounding single quotes ('') around the
        character or sequence of characters to protect. Inside single quotes, nothing

>         is interpreted, it's the efficient way to quote regular expressions.

Author: ok done
